{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15972672",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09439717",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79d7a985",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "\n",
    "device='cpu'\n",
    "import torch\n",
    "import numpy as np\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import torch\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "def set_seed(seed):\n",
    "    if seed is not None:\n",
    "        torch.manual_seed(seed)\n",
    "        torch.cuda.manual_seed_all(seed)\n",
    "        np.random.seed(seed)\n",
    "        random.seed(seed)\n",
    "        torch.backends.cudnn.deterministic = True\n",
    "        torch.backends.cudnn.benchmark = False\n",
    "        \n",
    "### BREEDING SIMULATOR\n",
    "class Genome:\n",
    "    def __init__(self, n_chr, n_loci, seed=None):\n",
    "        self.ploidy = 2\n",
    "        self.n_chr = n_chr\n",
    "        self.n_loci = n_loci\n",
    "        self.shape = (self.ploidy, self.n_chr, self.n_loci)\n",
    "\n",
    "\n",
    "class Population:\n",
    "    def __init__(self, genome, haplotypes, device=device):\n",
    "        self.genome = genome\n",
    "        self.device = device\n",
    "        self.phenotypes = None\n",
    "        self.bvs = None\n",
    "        self.haplotypes = haplotypes.to(device)\n",
    "        self.dosages = haplotypes.sum(dim=1).float().to(device)\n",
    "        self.size = haplotypes.shape[0]\n",
    "\n",
    "class Trait:\n",
    "    def __init__(self, genome, founder_population, target_mean, target_variance, device=device, seed=None):\n",
    "        set_seed(seed)\n",
    "        self.target_mean = target_mean\n",
    "        self.target_variance = target_variance\n",
    "        self.device = device\n",
    "        \n",
    "        random_effects = torch.randn(genome.n_chr, genome.n_loci, device=self.device)\n",
    "        random_effects -= random_effects.mean()\n",
    "        \n",
    "        founder_scores = torch.einsum('kl,hkl->h', random_effects, founder_population.dosages).to(device)\n",
    "        founder_mean, founder_var = founder_scores.mean(), founder_scores.var()\n",
    "        \n",
    "        scaling_factors = torch.sqrt(self.target_variance / founder_var)\n",
    "        self.scaling_factors = scaling_factors\n",
    "        random_effects *= scaling_factors\n",
    "        self.effects = random_effects\n",
    "        self.intercept = founder_mean - target_mean\n",
    "\n",
    "class SimParams:\n",
    "    def __init__(self, founder_pop, config):\n",
    "        self.G = Genome(config['n_chr'], config['n_loci'])\n",
    "        self.T = Trait(self.G, founder_pop, config['target_mean'], config['target_variance'], seed=config['seed'])\n",
    "        self.h2 = config['h2']\n",
    "        self.reps = config['reps']\n",
    "        self.pop_size = config['pop_size']\n",
    "        self.max_generations = config['start_gen']  # Initialize with start_gen\n",
    "        self.end_gen = config['end_gen']  # Store end_gen for reference\n",
    "        self.founder_pop = founder_pop\n",
    "        self.seed = config['seed']\n",
    "        if self.seed is not None:\n",
    "            set_seed(self.seed)\n",
    "\n",
    "    def update_max_generations(self, new_max_gen):\n",
    "        self.max_generations = min(new_max_gen, self.end_gen)\n",
    "\n",
    "def scale_values(x, from_range=(-1, 1), to_range=(0.05, 0.95)):\n",
    "    # Ensure input is a PyTorch tensor\n",
    "    if not isinstance(x, torch.Tensor):\n",
    "        x = torch.tensor(x, dtype=torch.float32)\n",
    "    \n",
    "    # Unpack the ranges\n",
    "    from_min, from_max = from_range\n",
    "    to_min, to_max = to_range\n",
    "    \n",
    "    # Scale the values\n",
    "    scaled = (x - from_min) / (from_max - from_min) * (to_max - to_min) + to_min\n",
    "    \n",
    "    return scaled\n",
    "\n",
    "\n",
    "## ACTIONS\n",
    "\n",
    "def calculate_breeding_value(population, trait, device=device):\n",
    "    return torch.einsum('hjk,jk->h', population.dosages, trait.effects).to(device)\n",
    "\n",
    "def truncation_selection(population, trait, top_percent):\n",
    "    return torch.topk(population.phenotypes, top_percent).indices.to(device)\n",
    "\n",
    "\n",
    "\n",
    "def bv(P,T):\n",
    "    P.breeding_values = calculate_breeding_value(P.dosages,T.effects)\n",
    "\n",
    "def create_pop(G, haplotypes):\n",
    "    return Population(G, haplotypes=haplotypes)\n",
    "\n",
    "def create_random_pop(G, pop_size):\n",
    "    return torch.randint(0, 2, (pop_size, *G.shape), device=device)\n",
    "\n",
    "def update_pop(population, haplotype_pop_tensor):\n",
    "    population.haplotypes = haplotype_pop_tensor\n",
    "    population.dosages = haplotype_pop_tensor.sum(dim=1).float()\n",
    "    return population\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# meiosis\n",
    "def recombine(parent_haplo_tensor, recombination_rate=0.1):\n",
    "    num_individuals, ploidy, num_chromosomes, num_loci = parent_haplo_tensor.shape\n",
    "    # Generate crossover masks\n",
    "    maternal, paternal = parent_haplo_tensor[:,0,:,:],parent_haplo_tensor[:,1,:,:],\n",
    "    crossovers = torch.bernoulli(torch.full((num_individuals, num_chromosomes, num_loci), recombination_rate, device=device))\n",
    "#     crossovers = torch.rand((num_individuals, num_chromosomes, num_loci), device=device) < recombination_rate\n",
    "    progeny = maternal * torch.logical_not(crossovers) + paternal * crossovers\n",
    "    return progeny\n",
    "\n",
    "def breed(mother_tensor, father_tensor, recombination_rate=0.1):\n",
    "    eggs = recombine(mother_tensor,recombination_rate)\n",
    "    pollens = recombine(father_tensor,recombination_rate)\n",
    "    return torch.stack((eggs,pollens), dim=1)\n",
    "\n",
    "\n",
    "\n",
    "def create_progeny(mother_gametes, father_gametes, reps=1, device=device):\n",
    "    # Repeat gametes along a new dimension for replications\n",
    "    mother_gametes = mother_gametes.unsqueeze(1).repeat(1, reps, 1, 1, 1)\n",
    "    father_gametes = father_gametes.unsqueeze(1).repeat(1, reps, 1, 1, 1)\n",
    "\n",
    "    # Shuffle indices for all replications at once\n",
    "    shuffled_mother_indices = torch.randperm(mother_gametes.shape[0], device=device)\n",
    "    shuffled_father_indices = torch.randperm(father_gametes.shape[0], device=device)\n",
    "    shuffled_mother_indices = shuffled_mother_indices.unsqueeze(1).repeat(1, reps)\n",
    "    shuffled_father_indices = shuffled_father_indices.unsqueeze(1).repeat(1, reps)\n",
    "\n",
    "    # Gather shuffled gametes and stack\n",
    "    mother_gametes = torch.gather(mother_gametes, 0, shuffled_mother_indices.unsqueeze(-1).unsqueeze(-1).unsqueeze(-1).expand(-1, -1, ploidy, num_chromosomes, num_loci))\n",
    "    father_gametes = torch.gather(father_gametes, 0, shuffled_father_indices.unsqueeze(-1).unsqueeze(-1).unsqueeze(-1).expand(-1, -1, ploidy, num_chromosomes, num_loci))\n",
    "    progeny_haplotypes = torch.stack((mother_gametes, father_gametes), dim=2)\n",
    "    return progeny_haplotypes.reshape(-1, 2, num_chromosomes, num_loci)\n",
    "\n",
    "\n",
    "def random_crosses(parent_population, total_crosses, device=device, seed=None):\n",
    "    if seed is not None:\n",
    "        torch.manual_seed(seed)\n",
    "    \n",
    "    num_parents = parent_population.shape[0]\n",
    "    ploidy, num_chromosomes, num_loci = parent_population.shape[1:]\n",
    "    \n",
    "    # Randomly select parents for each cross\n",
    "    parent_indices = torch.randint(0, num_parents, (total_crosses, 2), device=device)\n",
    "    \n",
    "    # Select the parent haplotypes\n",
    "    mothers = parent_population[parent_indices[:, 0]]\n",
    "    fathers = parent_population[parent_indices[:, 1]]\n",
    "    \n",
    "    # Perform recombination for both parents\n",
    "    mother_gametes = recombine(mothers)\n",
    "    father_gametes = recombine(fathers)\n",
    "    \n",
    "    # Combine gametes to create progeny\n",
    "    progeny = torch.stack((mother_gametes, father_gametes), dim=1)\n",
    "    \n",
    "    return progeny\n",
    "\n",
    "def phenotype(population, trait, h2):\n",
    "    breeding_values = calculate_breeding_value(population, trait)\n",
    "    population.breeding_values = breeding_values\n",
    "    population.genetic_var = breeding_values.var()\n",
    "    if breeding_values.var() == 0:\n",
    "        environmental_variance = 0\n",
    "    else:\n",
    "        environmental_variance = (1 - h2) / h2 * breeding_values.var()\n",
    "\n",
    "    # Check if environmental_variance is zero before applying torch.sqrt and .clone()\n",
    "    if environmental_variance == 0:\n",
    "        environmental_noise = torch.zeros(breeding_values.shape, device=device)\n",
    "    else:\n",
    "        environmental_noise = torch.randn(breeding_values.shape, device=device) * torch.sqrt(environmental_variance).detach()\n",
    "    population.phenotypes = breeding_values + environmental_noise\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6067f4bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "859be8a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
