{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7570e93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#x01_populationStatistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d9e0107f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import uuid\n",
    "import pdb\n",
    "import torch\n",
    "from matplotlib.animation import FuncAnimation\n",
    "import torch\n",
    "import torch.optim as optim\n",
    "import torch.nn.functional as F\n",
    "import torch.nn as nn\n",
    "import gymnasium as gym\n",
    "import numpy as np\n",
    "device='cpu'\n",
    "\n",
    "### BREEDING SIMULATOR\n",
    "class Genome:\n",
    "    def __init__(self, n_chr, n_loci):\n",
    "        self.ploidy = 2\n",
    "        self.n_chr = n_chr\n",
    "        self.n_loci = n_loci\n",
    "        self.shape = (self.ploidy, self.n_chr, self.n_loci)\n",
    "\n",
    "class Population:\n",
    "    def __init__(self, genome, haplotypes, device=device):\n",
    "        self.genome = genome\n",
    "        self.device = device\n",
    "        self.phenotypes = None\n",
    "        self.bvs = None\n",
    "        self.haplotypes = haplotypes.to(device)\n",
    "        self.dosages = haplotypes.sum(dim=1).float().to(device)\n",
    "        self.size = haplotypes.shape[0]\n",
    "\n",
    "class Trait:\n",
    "    def __init__(self, genome, founder_population, target_mean, target_variance, device=device):\n",
    "        self.target_mean = target_mean\n",
    "        self.target_variance = target_variance\n",
    "        self.device = device\n",
    "        random_effects = torch.randn(genome.n_chr, genome.n_loci, device=self.device)\n",
    "        random_effects -= random_effects.mean()\n",
    "        founder_scores = torch.einsum('kl,hkl->h', random_effects, founder_population.dosages).to(device)\n",
    "        founder_mean, founder_var = founder_scores.mean(), founder_scores.var()\n",
    "        scaling_factors = torch.sqrt(self.target_variance / founder_var)\n",
    "        self.scaling_factors = scaling_factors\n",
    "        random_effects *= scaling_factors\n",
    "        self.effects = random_effects\n",
    "        self.intercept = founder_mean - target_mean\n",
    "\n",
    "\n",
    "def calculate_breeding_value(population, trait, device=device):\n",
    "    return torch.einsum('hjk,jk->h', population.dosages, trait.effects).to(device)\n",
    "\n",
    "def truncation_selection(population, trait, top_percent):\n",
    "    return torch.topk(population.phenotypes, top_percent).indices.to(device)\n",
    "\n",
    "# meiosis\n",
    "def recombine(parent_haplo_tensor, recombination_rate=0.1):\n",
    "    num_individuals, ploidy, num_chromosomes, num_loci = parent_haplo_tensor.shape\n",
    "    # Generate crossover masks\n",
    "    maternal, paternal = parent_haplo_tensor[:,0,:,:], parent_haplo_tensor[:,1,:,:]\n",
    "    crossovers = torch.bernoulli(torch.full((num_individuals, num_chromosomes, num_loci), recombination_rate, device=device))\n",
    "    progeny = maternal * (1 - crossovers) + paternal * crossovers\n",
    "    return progeny\n",
    "\n",
    "\n",
    "\n",
    "def create_random_pop(G, pop_size):\n",
    "    return torch.randint(0, 2, (pop_size, *G.shape), device=device)\n",
    "\n",
    "def truncation_selection(population, trait, top_percent):\n",
    "    return torch.topk(population.phenotypes, top_percent).indices.to(device)\n",
    "\n",
    "# meiosis\n",
    "def recombine(parent_haplo_tensor, recombination_rate=0.1):\n",
    "    num_individuals, ploidy, num_chromosomes, num_loci = parent_haplo_tensor.shape\n",
    "    # Generate crossover masks\n",
    "    maternal, paternal = parent_haplo_tensor[:,0,:,:], parent_haplo_tensor[:,1,:,:]\n",
    "    crossovers = torch.bernoulli(torch.full((num_individuals, num_chromosomes, num_loci), recombination_rate, device=device))\n",
    "    progeny = maternal * (1 - crossovers) + paternal * crossovers\n",
    "    return progeny\n",
    "\n",
    "def phenotype(population, trait, h2):\n",
    "    breeding_values = calculate_breeding_value(population, trait)\n",
    "    population.breeding_values = breeding_values\n",
    "    if breeding_values.var() == 0:\n",
    "        environmental_variance = 0\n",
    "    else:\n",
    "        environmental_variance = (1 - h2) / h2 * breeding_values.var()\n",
    "\n",
    "    # Check if environmental_variance is zero before applying torch.sqrt and .clone()\n",
    "    if environmental_variance == 0:\n",
    "        environmental_noise = torch.zeros(breeding_values.shape, device=device)\n",
    "    else:\n",
    "        environmental_noise = torch.randn(breeding_values.shape, device=device) * torch.sqrt(environmental_variance).detach()\n",
    "\n",
    "    population.breeding_values = breeding_values\n",
    "    population.phenotypes = breeding_values + environmental_noise\n",
    "\n",
    "    return population.phenotypes.max()\n",
    "\n",
    "def create_random_pop(G, pop_size):\n",
    "    return torch.randint(0, 2, (pop_size, *G.shape), device=device)\n",
    "\n",
    "def update_pop(population, haplotype_pop_tensor):\n",
    "    population.haplotypes = haplotype_pop_tensor\n",
    "    population.dosages = haplotype_pop_tensor.sum(dim=1).float()\n",
    "    return population\n",
    "\n",
    "# meiosis\n",
    "def recombine(parent_haplo_tensor, recombination_rate=0.1):\n",
    "    num_individuals, ploidy, num_chromosomes, num_loci = parent_haplo_tensor.shape\n",
    "    # Generate crossover masks\n",
    "    maternal, paternal = parent_haplo_tensor[:,0,:,:],parent_haplo_tensor[:,1,:,:],\n",
    "    crossovers = torch.bernoulli(torch.full((num_individuals, num_chromosomes, num_loci), recombination_rate, device=device))\n",
    "#     crossovers = torch.rand((num_individuals, num_chromosomes, num_loci), device=device) < recombination_rate\n",
    "    progeny = maternal * torch.logical_not(crossovers) + paternal * crossovers\n",
    "    return progeny\n",
    "\n",
    "def breed(mother_tensor, father_tensor, recombination_rate=0.1):\n",
    "    eggs = recombine(mother_tensor,recombination_rate)\n",
    "    pollens = recombine(father_tensor,recombination_rate)\n",
    "    return torch.stack((eggs,pollens), dim=1)\n",
    "\n",
    "def create_pop(G, haplotypes):\n",
    "    return Population(G, haplotypes=haplotypes)\n",
    "\n",
    "def bv(P,T):\n",
    "    P.breeding_values = calculate_breeding_value(P.dosages,T.effects)\n",
    "\n",
    "def create_progeny(mother_gametes, father_gametes, reps=1, device=device):\n",
    "    progeny = []\n",
    "    for _ in range(reps):\n",
    "        # Randomly shuffle the gametes from each parent\n",
    "        shuffled_mother_indices = torch.randperm(mother_gametes.shape[0], device=device)\n",
    "        shuffled_father_indices = torch.randperm(father_gametes.shape[0], device=device)\n",
    "\n",
    "        # Select the shuffled gametes\n",
    "        mother_gametes = mother_gametes[shuffled_mother_indices]\n",
    "        father_gametes = father_gametes[shuffled_father_indices]\n",
    "\n",
    "        # Stack the gametes to create progeny haplotypes\n",
    "        progeny_haplotypes = torch.stack((mother_gametes, father_gametes), dim=1)\n",
    "        progeny.append(progeny_haplotypes)\n",
    "    return torch.vstack(progeny)\n",
    "\n",
    "\n",
    "def random_crosses(parent_population, total_crosses, device=device):\n",
    "    num_parents = parent_population.shape[0]\n",
    "    ploidy, num_chromosomes, num_loci = parent_population.shape[1:]\n",
    "    \n",
    "    # Randomly select parents for each cross\n",
    "    parent_indices = torch.randint(0, num_parents, (total_crosses, 2), device=device)\n",
    "    \n",
    "    # Select the parent haplotypes\n",
    "    mothers = parent_population[parent_indices[:, 0]]\n",
    "    fathers = parent_population[parent_indices[:, 1]]\n",
    "    \n",
    "    # Perform recombination for both parents\n",
    "    mother_gametes = recombine(mothers)\n",
    "    father_gametes = recombine(fathers)\n",
    "    \n",
    "    # Combine gametes to create progeny\n",
    "    progeny = torch.stack((mother_gametes, father_gametes), dim=1)\n",
    "    \n",
    "    return progeny\n",
    "\n",
    "class SimParams:\n",
    "    def __init__(self,G,T,h2,reps,pop_size,max_generations,founder_pop):\n",
    "        self.G = G\n",
    "        self.T = T\n",
    "        self.h2 = h2\n",
    "        self.reps = reps\n",
    "        self.pop_size = pop_size\n",
    "        self.max_generations = max_generations\n",
    "        self.founder_pop = founder_pop\n",
    "\n",
    "#RL ENVIRONMENT\n",
    "class BreedingEnvironment(gym.Env):\n",
    "    def __init__(self, SP):\n",
    "        super(BreedingEnvironment, self).__init__()\n",
    "        self.SP = SP\n",
    "        self.current_generation = 0\n",
    "        self.max_generations = SP.max_generations\n",
    "\n",
    "\n",
    "        # Define action and observation space\n",
    "        self.action_space = gym.spaces.Discrete(5)  # 0, 1, 2, 3, or 4\n",
    "        self.action_history = []\n",
    "\n",
    "        self.observation_space = gym.spaces.Dict({\n",
    "            \"population\": gym.spaces.Box(low=0, high=1, shape=(200, 2, 1, 200), dtype=np.int32),\n",
    "            \"generation\": gym.spaces.Box(low=0, high=1, shape=(1,), dtype=np.float32)\n",
    "        })\n",
    "\n",
    "        self._action_to_direction = {0:(200,1),\n",
    "               1:(100,2),\n",
    "               2:(50,4),\n",
    "               3:(25,8),\n",
    "               4:(5,40),}\n",
    "\n",
    "    def _get_obs(self):\n",
    "        return {\n",
    "            \"population\": self.population.haplotypes.cpu(),\n",
    "            \"generation\": torch.tensor([self.generation / self.SP.max_generations], dtype=torch.float32).cpu()\n",
    "        }\n",
    "\n",
    "    def _get_info(self):\n",
    "        return {\n",
    "            \"phenotype\": self.phenotype.cpu().item(),\n",
    "            \"genetic_variance\": self.population.breeding_values.var().cpu().item()\n",
    "        }\n",
    "\n",
    "    def reset(self, seed=None, options=None):\n",
    "        super().reset(seed=seed)\n",
    "        self.population = self.SP.founder_pop\n",
    "        self.phenotype = phenotype(self.population, self.SP.T, self.SP.h2)\n",
    "        self.generation = 0\n",
    "\n",
    "        observation = self._get_obs()\n",
    "        info = self._get_info()\n",
    "        self.action_history = []\n",
    "        return observation, info\n",
    "\n",
    "    def step(self, action):\n",
    "        # Map the action (element of {0,1,2,3}) to the direction we walk in\n",
    "#         print(action)\n",
    "        total_parents, total_crosses = self._action_to_direction[int(action)] # top parents, number crosses per\n",
    "\n",
    "        top_k = torch.topk(self.population.phenotypes, total_parents).indices\n",
    "        selected = self.population.haplotypes[top_k]\n",
    "\n",
    "        # Breeding\n",
    "        m = recombine(selected)  # Mother gametes\n",
    "        f = recombine(selected)  # Father gametes\n",
    "        progeny = create_progeny(m, f, reps=total_crosses)\n",
    "\n",
    "        # Create new population from progeny\n",
    "        self.population = create_pop(self.SP.G, progeny)\n",
    "        self.phenotype = phenotype(self.population, self.SP.T, self.SP.h2)\n",
    "\n",
    "\n",
    "        self.generation+=1\n",
    "        observation = self._get_obs()\n",
    "        info = self._get_info()\n",
    "        terminated = self.generation >= self.SP.max_generations\n",
    "        truncated = False\n",
    "        reward = self.phenotype\n",
    "        self.action_history.append((action))\n",
    "        self.current_generation += 1\n",
    "        done = self.current_generation >= self.max_generations\n",
    "\n",
    "        if done:\n",
    "            info['final_generation'] = {\n",
    "                'phenotype': self.phenotype,\n",
    "                'genetic_variance': self.population.breeding_values.var().item()\n",
    "            }\n",
    "        return observation, reward, terminated, truncated, info\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e2cf7ea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def phenotype(population, trait, h2):\n",
    "    breeding_values = calculate_breeding_value(population, trait)\n",
    "    population.breeding_values = breeding_values\n",
    "    population.genetic_var = breeding_values.var()\n",
    "    if breeding_values.var() == 0:\n",
    "        environmental_variance = 0\n",
    "    else:\n",
    "        environmental_variance = (1 - h2) / h2 * breeding_values.var()\n",
    "\n",
    "    # Check if environmental_variance is zero before applying torch.sqrt and .clone()\n",
    "    if environmental_variance == 0:\n",
    "        environmental_noise = torch.zeros(breeding_values.shape, device=device)\n",
    "    else:\n",
    "        environmental_noise = torch.randn(breeding_values.shape, device=device) * torch.sqrt(environmental_variance).detach()\n",
    "    population.phenotypes = breeding_values + environmental_noise\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a40461dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8bb0a3ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using cpu device\n",
      "-----------------------------\n",
      "| time/              |      |\n",
      "|    fps             | 141  |\n",
      "|    iterations      | 1    |\n",
      "|    time_elapsed    | 14   |\n",
      "|    total_timesteps | 2048 |\n",
      "-----------------------------\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<stable_baselines3.ppo.ppo.PPO at 0x7f3db6872c50>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gymnasium as gym\n",
    "import numpy as np\n",
    "import torch\n",
    "from gymnasium import spaces\n",
    "from stable_baselines3 import PPO\n",
    "from stable_baselines3.common.vec_env import DummyVecEnv\n",
    "from stable_baselines3.common.env_checker import check_env\n",
    "\n",
    "class SelectionIntensityEnvironment(gym.Env):\n",
    "    def __init__(self, SP):\n",
    "        super(SelectionIntensityEnvironment, self).__init__()\n",
    "        self.SP = SP\n",
    "        self.current_generation = 0\n",
    "        self.max_generations = SP.max_generations\n",
    "        # Define action space\n",
    "        self.action_space = spaces.Box(\n",
    "            low=np.array([0.1]), \n",
    "            high=np.array([1.0]), \n",
    "            dtype=np.float32\n",
    "        )\n",
    "        # Define observation space\n",
    "        self.observation_space = gym.spaces.Dict({\n",
    "            \"population\": gym.spaces.Box(low=0, high=1, shape=(self.SP.pop_size, 2, self.SP.G.n_chr, self.SP.G.n_loci), dtype=np.int32),\n",
    "            \"generation\": gym.spaces.Box(low=0, high=1, shape=(1,), dtype=np.float32)\n",
    "        })\n",
    "        # Logging attributes\n",
    "        self.action_values = []\n",
    "        self.genetic_variance = []\n",
    "        self.max_breeding_values = []\n",
    "        self.final_generations = []\n",
    "        self.episode_count = 0\n",
    "        \n",
    "    def _get_obs(self):\n",
    "        population = self.population.haplotypes.cpu().numpy().astype(np.int32)\n",
    "        generation = np.array([self.current_generation / self.SP.max_generations], dtype=np.float32)\n",
    "        return {\n",
    "            \"population\": population,\n",
    "            \"generation\": generation\n",
    "        }\n",
    "\n",
    "    def _get_info(self):\n",
    "        return {\n",
    "            \"phenotype\": self.population.phenotypes.max().cpu().item(),\n",
    "            \"genetic_variance\": self.population.breeding_values.var().cpu().item()\n",
    "        }\n",
    "\n",
    "    def reset(self, seed=None, options=None):\n",
    "        super().reset(seed=seed)\n",
    "        self.population = self.SP.founder_pop\n",
    "        self.phenotype = phenotype(self.population, self.SP.T, self.SP.h2)\n",
    "\n",
    "        observation = self._get_obs()\n",
    "        info = self._get_info()\n",
    "        return observation, info\n",
    "\n",
    "    def step(self, action):\n",
    "        action_scalar = action.item() if isinstance(action, np.ndarray) else action\n",
    "        selected = torch.topk(self.population.phenotypes, int(action_scalar * self.population.size)).indices\n",
    "        new_pop = random_crosses(self.population.haplotypes[selected], self.SP.pop_size)\n",
    "        \n",
    "        self.current_generation += 1\n",
    "        observation = self._get_obs()\n",
    "        info = self._get_info()\n",
    "        terminated = self.current_generation >= self.SP.max_generations or self.population.genetic_var < .005\n",
    "        reward = float(self.population.breeding_values.max())\n",
    "        \n",
    "        \n",
    "        self.action_values.append((self.current_generation, action_scalar))\n",
    "        # Log genetic variance and max breeding value at the end of each generation\n",
    "        if terminated:\n",
    "            self.genetic_variance.append(self.population.genetic_var)\n",
    "            self.max_breeding_values.append(float(self.population.breeding_values.max()))\n",
    "            self.final_generations.append(self.current_generation)\n",
    "            self.episode_count += 1\n",
    "        \n",
    "        return observation, reward, bool(terminated), False, info\n",
    "\n",
    "# Example usage\n",
    "n_chr = 1\n",
    "n_loci = 222\n",
    "founder_pop_size = 333\n",
    "h2 = 1\n",
    "reps = 1\n",
    "max_generations = 20\n",
    "G = Genome(n_chr, n_loci)\n",
    "founder_pop = create_pop(G, create_random_pop(G, 5))\n",
    "founder_pop = random_crosses(founder_pop.haplotypes, founder_pop_size)\n",
    "founder_pop = create_pop(G, founder_pop)\n",
    "T = Trait(G, founder_pop, target_mean=0.0, target_variance=1.0)\n",
    "SP = SimParams(G, T, h2, reps, founder_pop_size, max_generations, founder_pop)\n",
    "\n",
    "env = SelectionIntensityEnvironment(SP)\n",
    "check_env(env)\n",
    "\n",
    "vec_env = DummyVecEnv([lambda: env])\n",
    "model = PPO(\"MultiInputPolicy\", vec_env, verbose=1)\n",
    "total_timesteps = 1000\n",
    "model.learn(total_timesteps=total_timesteps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad71eef4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
