{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df15fd50",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d3183a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp trait"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ace19fe9",
   "metadata": {},
   "source": [
    "## Trait\n",
    "> contains logic for quantitative traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ec1afd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#|export\n",
    "\n",
    "from chewc.core import *\n",
    "import torch\n",
    "import attr\n",
    "from typing import Tuple, Optional, List\n",
    "from fastcore.test import *\n",
    "import matplotlib.pyplot as plt\n",
    "import torch.nn as nn\n",
    "import pdb\n",
    "\n",
    "def select_qtl_loci(num_qtl_per_chromosome: int, genome: Genome) -> torch.Tensor:\n",
    "    \"\"\"\n",
    "    Randomly selects loci to be QTLs on each chromosome.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "    num_qtl_per_chromosome (int): Number of QTLs to select per chromosome.\n",
    "    genome (Genome): Genome object containing the chromosome structure.\n",
    "    device (str): Device to create the tensor on ('cpu' or 'cuda'). Defaults to 'cpu'.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    torch.Tensor: A boolean tensor indicating which loci are QTLs. \n",
    "                  Shape: (number_chromosomes, loci_per_chromosome)\n",
    "    \"\"\"\n",
    "    \n",
    "    assert num_qtl_per_chromosome <= genome.n_loci_per_chromosome, \"too many qtls for this trait given your Genome object\"\n",
    "    assert num_qtl_per_chromosome > 0, \"you need at least 1 QTL per chromosome\"\n",
    "    \n",
    "    qtl_indices = []\n",
    "    for i in range(genome.n_chromosomes):\n",
    "        # Randomly sample indices for QTLs on the current chromosome\n",
    "        chromosome_indices = torch.randperm(genome.n_loci_per_chromosome)[:num_qtl_per_chromosome]\n",
    "        \n",
    "        # Create a boolean tensor for the current chromosome, marking QTL positions as True\n",
    "        chromosome_qtl_flags = torch.zeros(genome.n_loci_per_chromosome, dtype=torch.bool)\n",
    "        chromosome_qtl_flags[chromosome_indices] = True\n",
    "        \n",
    "        qtl_indices.append(chromosome_qtl_flags)\n",
    "    \n",
    "    return torch.stack(qtl_indices)\n",
    "\n",
    "class TraitA(nn.Module):\n",
    "    def __init__(self, genome: Genome, effects: torch.Tensor, intercept: float):\n",
    "        \"\"\"\n",
    "        Initialize the TraitA model.\n",
    "\n",
    "        Args:\n",
    "            genome (Genome): The genome object containing the genomic architecture.\n",
    "            effects (torch.Tensor): The QTL effects tensor. Shape: (n_chromosomes, n_loci_per_chromosome).\n",
    "            intercept (float): The intercept value for the trait model.\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "        self.genome = genome\n",
    "        self.effects = effects.to(genome.device)\n",
    "        self.intercept = intercept\n",
    "\n",
    "    def forward(self, dosages: torch.Tensor, h2: Optional[float] = None, varE: Optional[float] = None) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Forward pass to calculate the breeding values with optional environmental noise.\n",
    "\n",
    "        Args:\n",
    "            dosages (torch.Tensor): The dosage tensor. Shape: (population_size, n_chromosomes, n_loci_per_chromosome).\n",
    "            h2 (float, optional): Heritability of the trait. If provided, environmental noise is added based on h2.\n",
    "            varE (float, optional): Environmental variance. If provided, environmental noise is added based on varE.\n",
    "\n",
    "        Returns:\n",
    "            torch.Tensor: The calculated breeding values with optional environmental noise.\n",
    "        \"\"\"\n",
    "        # Calculate breeding values\n",
    "        breeding_values = torch.einsum('ijk,jk->i', dosages, self.effects) + self.intercept\n",
    "\n",
    "        \n",
    "        if varE is not None:\n",
    "            # Add environmental noise based on environmental variance\n",
    "            env_noise = torch.randn_like(breeding_values) * torch.sqrt(torch.tensor(varE))\n",
    "            breeding_values += env_noise\n",
    "        elif h2 is not None:\n",
    "            # Calculate environmental variance from heritability\n",
    "            varG = breeding_values.var(unbiased=False)\n",
    "            varE = varG * (1 - h2) / h2\n",
    "            env_noise = torch.randn_like(breeding_values) * torch.sqrt(varE)\n",
    "            breeding_values += env_noise\n",
    "\n",
    "        return breeding_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f4b77c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def corr_traits(genome, init_pop,  target_means, target_vars, correlation_matrix):\n",
    "    \n",
    "    n_chr, n_loci = genome.genetic_map.shape\n",
    "    n_traits = target_means.shape[0]\n",
    "    corA = torch.tensor(correlation_matrix)\n",
    "    L = torch.linalg.cholesky(corA)\n",
    "\n",
    "    # Sample initial additive effects from a standard normal distribution\n",
    "    uncorrelated_effects = torch.randn(n_chr, n_loci, n_traits)\n",
    "\n",
    "    # Reshape for proper multiplication with Cholesky factor\n",
    "    uncorrelated_effects = uncorrelated_effects.reshape(n_chr * n_loci, n_traits)\n",
    "\n",
    "    # Introduce correlation FIRST\n",
    "    correlated_effects = torch.matmul(L, uncorrelated_effects.T).T\n",
    "    correlated_effects = correlated_effects.reshape(n_chr, n_loci, n_traits)\n",
    "\n",
    "    # Calculate unscaled breeding values using CORRELATED effects\n",
    "    unscaled_bvs = torch.einsum('ijk,lij->lk', correlated_effects, init_pop)\n",
    "    unscaled_var = unscaled_bvs.var(dim=0)\n",
    "    unscaled_mean = unscaled_bvs.mean(dim=0)\n",
    "    trait_intercepts = []\n",
    "\n",
    "    # Scale correlated effects and calculate intercepts\n",
    "    for i in range(n_traits):\n",
    "        scaling_factor = torch.sqrt(target_vars[i] / unscaled_var[i])\n",
    "        correlated_effects[:, :, i] *= scaling_factor  # Scale the CORRELATED effects\n",
    "        trait_intercepts.append(target_means[i] - (unscaled_mean[i] * scaling_factor))\n",
    "\n",
    "    # Now we have the additive marker effects and intercepts to calculate breeding values\n",
    "    trait_intercepts = torch.tensor(trait_intercepts)\n",
    "#     scaled_bvs = torch.einsum('ijk,lij->lk', correlated_effects, init_pop)\n",
    "    \n",
    "    \n",
    "    Traits = []\n",
    "    for i in range(len(target_means)):\n",
    "        Traits.append(TraitA(genome, correlated_effects[:,:,i], trait_intercepts[i]))\n",
    "    return Traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a757cf4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created genetic map\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "ploidy = 2\n",
    "n_chr = 10\n",
    "n_loci = 100\n",
    "n_Ind = 333\n",
    "g = Genome(ploidy, n_chr, n_loci)\n",
    "population = Population()\n",
    "population.create_random_founder_population(g, n_founders=n_Ind)\n",
    "init_pop = population.get_dosages().float()  # gets allele dosage for calculating trait values\n",
    "\n",
    "# multi_traits\n",
    "target_means = torch.tensor([0, 5, 20])\n",
    "target_vars = torch.tensor([1, 1, 0.5])  # Note: I'm assuming you want a variance of 1 for the second trait\n",
    "correlation_values = [\n",
    "        [1.0, 0.2, 0.58],\n",
    "        [0.2, 1.0, -0.37],\n",
    "        [0.58, -0.37, 1.0],\n",
    "    ]\n",
    "\n",
    "\n",
    "correlated_traits = corr_traits(g, init_pop, target_means, target_vars, correlation_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f836f5b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
