{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df15fd50",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d3183a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp trait"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ad173a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from burbankai.core import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ace19fe9",
   "metadata": {},
   "source": [
    "## Trait\n",
    "> contains logic for quantitative traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8d86682",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "\n",
    "from burbankai.core import *\n",
    "\n",
    "import torch\n",
    "import attr\n",
    "\n",
    "def select_qtl_loci(num_qtl_per_chromosome: int, genome:Genome) -> torch.Tensor:\n",
    "    \"\"\"\n",
    "    Randomly selects loci to be QTLs on each chromosome.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "    num_qtl_per_chromosome (int): Number of QTLs to select per chromosome.\n",
    "    genome (Genome): Genome object containing the chromosome structure.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    torch.Tensor: A boolean tensor indicating which loci are QTLs. \n",
    "                  Shape: (number_chromosomes, loci_per_chromosome)\n",
    "    \"\"\"\n",
    "    qtl_indices = []\n",
    "    for i in range(genome.number_chromosomes):\n",
    "        # Randomly sample indices for QTLs on the current chromosome\n",
    "        chromosome_indices = torch.randperm(genome.loci_per_chromosome)[:num_qtl_per_chromosome]\n",
    "        \n",
    "        # Create a boolean tensor for the current chromosome, marking QTL positions as True\n",
    "        chromosome_qtl_flags = torch.zeros(genome.loci_per_chromosome, dtype=torch.bool)\n",
    "        chromosome_qtl_flags[chromosome_indices] = True\n",
    "        \n",
    "        qtl_indices.append(chromosome_qtl_flags)\n",
    "    \n",
    "    return torch.stack(qtl_indices)\n",
    "\n",
    "def generate_marker_effects(qtl_map: torch.Tensor, mean: float = 0.0, variance: float = 1.0) -> torch.Tensor:\n",
    "    \"\"\"\n",
    "    Generates random marker effects for QTLs, drawn from a normal distribution.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "    qtl_map (torch.Tensor): A boolean tensor indicating which loci are QTLs. \n",
    "                            Shape: (number_chromosomes, loci_per_chromosome)\n",
    "    mean (float): The mean of the normal distribution from which to draw effects. Defaults to 0.0.\n",
    "    variance (float): The variance of the normal distribution from which to draw effects. Defaults to 1.0.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    torch.Tensor: A tensor of marker effects. Shape: (number_chromosomes, loci_per_chromosome).\n",
    "                  Non-QTL loci will have an effect of 0.\n",
    "    \"\"\"\n",
    "    # Create a tensor of zeros with the same shape as the qtl_map\n",
    "    effects = torch.zeros_like(qtl_map, dtype=torch.float)\n",
    "    \n",
    "    # Determine the number of QTLs\n",
    "    num_qtl = qtl_map.sum().item()\n",
    "\n",
    "    # Sample random effects from a normal distribution\n",
    "    qtl_effects = torch.randn(num_qtl) * (variance ** 0.5) + mean\n",
    "    \n",
    "    # Assign the sampled effects to the QTL positions in the effects tensor\n",
    "    effects[qtl_map] = qtl_effects\n",
    "    \n",
    "    return effects\n",
    "def calculate_genetic_variance(founder_pop: torch.Tensor, marker_effects: torch.Tensor, genome: Genome) -> float:\n",
    "    \"\"\"\n",
    "    Calculates the additive genetic variance in the founder population.\n",
    "\n",
    "    Args:\n",
    "        founder_pop (torch.Tensor): Tensor of founder haplotypes. \n",
    "                                    Shape: (n_founders, ploidy, number_chromosomes, loci_per_chromosome)\n",
    "        marker_effects (torch.Tensor): Tensor of marker effects. \n",
    "                                       Shape: (number_chromosomes, loci_per_chromosome)\n",
    "        genome (Genome): The genome object.\n",
    "\n",
    "    Returns:\n",
    "        torch.Tensor: The additive genetic variance.\n",
    "    \"\"\"\n",
    "    # Convert to float for mean operation to work\n",
    "    founder_pop = founder_pop.float()\n",
    "    # Calculate allele frequencies in the founder population\n",
    "    allele_frequencies = founder_pop.mean(dim=(0, 1))  # Average across founders and ploidy\n",
    "\n",
    "    # Calculate the genetic value contributions of each locus\n",
    "    locus_contributions = 2 * allele_frequencies * (1 - allele_frequencies) * marker_effects**2\n",
    "\n",
    "    # Sum the contributions across all loci to get the total genetic variance\n",
    "    genetic_variance = locus_contributions.sum().item()\n",
    "\n",
    "    return genetic_variance\n",
    "\n",
    "def scale_marker_effects(marker_effects: torch.Tensor, \n",
    "                         initial_variance: float, \n",
    "                         desired_variance: float) -> torch.Tensor:\n",
    "    \"\"\"\n",
    "    Scales marker effects to achieve a desired genetic variance.\n",
    "\n",
    "    Args:\n",
    "        marker_effects (torch.Tensor): The initial marker effects.\n",
    "        initial_variance (float): The genetic variance with the initial effects.\n",
    "        desired_variance (float): The desired genetic variance.\n",
    "\n",
    "    Returns:\n",
    "        torch.Tensor: The scaled marker effects.\n",
    "    \"\"\"\n",
    "    scaling_factor = (desired_variance / initial_variance) ** 0.5\n",
    "    return marker_effects * scaling_factor\n",
    "\n",
    "#| hide\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1ea8898",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "@attr.s(auto_attribs=True)\n",
    "class TraitA:\n",
    "    \"\"\"\n",
    "    Represents a trait with only additive genetic effects.\n",
    "\n",
    "    Attributes:\n",
    "    ----------\n",
    "    qtl_map (torch.Tensor): A boolean tensor indicating which loci are QTLs. \n",
    "                           Shape: (number_chromosomes, loci_per_chromosome)\n",
    "    additive_effects (torch.Tensor): A tensor of additive effects for each QTL. \n",
    "                                  Shape: (number_chromosomes, loci_per_chromosome)\n",
    "    genome (Genome): The genome object.\n",
    "    founder_pop (torch.Tensor): Tensor of founder haplotypes. \n",
    "                                Shape: (n_founders, ploidy, number_chromosomes, loci_per_chromosome)\n",
    "    target_variance (float): The desired genetic variance for the trait.\n",
    "    intercept (float): The intercept value, calculated during initialization.\n",
    "    \n",
    "    Methods:\n",
    "    -------\n",
    "    calculate_genetic_value(genotypes: torch.Tensor) -> torch.Tensor:\n",
    "        Calculates the genetic value of individuals based on their genotypes.\n",
    "    _calculate_intercept() -> float:\n",
    "        Calculates the intercept based on the mean genetic value of the founder population.\n",
    "    _calculate_scaled_additive_dosages(genotypes: torch.Tensor) -> torch.Tensor:\n",
    "        Calculates the scaled additive genotype dosages.\n",
    "    _scale_effects() -> None:\n",
    "        Scales the additive effects to achieve the target genetic variance.\n",
    "    \"\"\"\n",
    "\n",
    "    qtl_map: torch.Tensor \n",
    "    additive_effects: torch.Tensor \n",
    "    genome: Genome \n",
    "    founder_pop: torch.Tensor \n",
    "    target_variance: float\n",
    "    target_mean: float\n",
    "    intercept: float = attr.ib(init=False)\n",
    "\n",
    "    def __attrs_post_init__(self):\n",
    "        \"\"\"\n",
    "        Calculate the intercept and scale the effects after initialization.\n",
    "        \"\"\"\n",
    "        self.intercept = self._calculate_intercept()\n",
    "        self._scale_effects() \n",
    "\n",
    "    def _calculate_intercept(self) -> float:\n",
    "        \"\"\"\n",
    "        Calculates the intercept based on the mean genetic value of the founder population.\n",
    "\n",
    "        Returns:\n",
    "            float: The intercept value.\n",
    "        \"\"\"\n",
    "        # Calculate the mean genetic value of the founder population (without scaling)\n",
    "        founder_genetic_values = (self.founder_pop.float() * self.additive_effects).sum(dim=(1, 2, 3))\n",
    "        mean_founder_gv = founder_genetic_values.mean().item()\n",
    "        return mean_founder_gv\n",
    "\n",
    "    def _calculate_scaled_additive_dosages(self, genotypes: torch.Tensor) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Calculates the scaled additive genotype dosages.\n",
    "\n",
    "        Args:\n",
    "            genotypes (torch.Tensor): A tensor representing the genotypes of individuals.\n",
    "                                     Shape: (n_individuals, ploidy, number_chromosomes, loci_per_chromosome).\n",
    "\n",
    "        Returns:\n",
    "            torch.Tensor: A tensor of scaled additive dosages. \n",
    "                         Shape: (n_individuals, ploidy, number_chromosomes, loci_per_chromosome).\n",
    "        \"\"\"\n",
    "        return (genotypes - self.genome.ploidy / 2) * (2 / self.genome.ploidy)\n",
    "\n",
    "    def _scale_effects(self) -> None:\n",
    "        \"\"\"\n",
    "        Scales the additive effects to achieve the target genetic variance and \n",
    "        calculates the intercept to achieve the target mean.\n",
    "        \"\"\"\n",
    "        # Calculate the initial genetic variance in the founder population\n",
    "        founder_gvs = self.calculate_genetic_value(self.founder_pop)\n",
    "        initial_variance = founder_gvs.var().item()\n",
    "\n",
    "        # Calculate the scaling factor\n",
    "        scaling_factor = (self.target_variance / initial_variance) ** 0.5\n",
    "\n",
    "        # Scale the additive effects\n",
    "        self.additive_effects = self.additive_effects * scaling_factor\n",
    "\n",
    "        # Recalculate the intercept after scaling\n",
    "        self.intercept = self.target_mean - self.calculate_genetic_value(self.founder_pop).mean().item()\n",
    "\n",
    "    def calculate_genetic_value(self, genotypes: torch.Tensor) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Calculates the genetic value of individuals given their genotypes.\n",
    "\n",
    "        Args:\n",
    "        ----\n",
    "        genotypes (torch.Tensor): A tensor representing the genotypes of individuals.\n",
    "                                 Shape: (n_individuals, ploidy, number_chromosomes, loci_per_chromosome).\n",
    "\n",
    "        Returns:\n",
    "        -------\n",
    "        torch.Tensor: A tensor of genetic values for each individual. Shape: (n_individuals,).\n",
    "        \"\"\"\n",
    "\n",
    "        # Calculate the scaled additive genotype dosages\n",
    "        scaled_dosages = self._calculate_scaled_additive_dosages(genotypes)\n",
    "\n",
    "        # Apply the additive effects to the scaled dosages, only at QTL positions\n",
    "        additive_genetic_values = (scaled_dosages * self.additive_effects).sum(dim=(1, 2, 3))\n",
    "\n",
    "        # Add the target mean to adjust the mean genetic value\n",
    "        return additive_genetic_values + self.target_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39f01e03",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a3d5fdd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fd478f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 2.9670262e-01, -1.0608990e+00, -1.3061762e-01, -6.8765855e-01,\n",
       "        7.5890046e-01, -4.3069351e-01, -4.3168354e-01,  4.3027699e-01,\n",
       "        1.6136456e+00,  9.3925142e-01,  1.0389560e+00,  2.7477545e-01,\n",
       "       -1.2480235e-01,  6.0117507e-01, -4.9601614e-01,  2.9064101e-01,\n",
       "        6.8575621e-02,  7.5689793e-02, -5.0014019e-02, -5.6168795e-02,\n",
       "        1.1408436e-01,  3.4718472e-01,  1.4926574e+00, -5.4286718e-02,\n",
       "       -1.0044148e+00,  2.4692625e-01, -6.6761780e-01, -2.9217243e-02,\n",
       "        4.7739625e-02, -5.5097139e-01, -1.0332935e+00,  1.2361276e-01,\n",
       "        1.6606410e+00, -4.0385544e-01,  1.7578388e+00,  2.3948038e-01,\n",
       "       -2.6294327e-01,  8.5001546e-01, -8.8876772e-01,  4.2174608e-01,\n",
       "        5.3662062e-04,  6.6406792e-01,  1.1790458e+00,  6.0373819e-01,\n",
       "       -8.0300570e-01,  3.7246078e-01,  1.4497381e-01,  9.3769932e-01,\n",
       "        4.4031417e-01,  1.0456161e+00], dtype=float32)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "749ea0ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
