{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "{}\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df15fd50",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d3183a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp trait"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ace19fe9",
   "metadata": {},
   "source": [
    "## Trait\n",
    "> contains logic for quantitative traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ec1afd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#|export\n",
    "\n",
    "from chewc.core import *\n",
    "import torch\n",
    "import attr\n",
    "from typing import Tuple, Optional, List, Union\n",
    "from fastcore.test import *\n",
    "import matplotlib.pyplot as plt\n",
    "import torch.nn as nn\n",
    "import pdb\n",
    "\n",
    "\n",
    "def select_qtl_loci(num_qtl_per_chromosome: int, genome: Genome) -> torch.Tensor:\n",
    "    \"\"\"\n",
    "    Randomly selects loci to be QTLs on each chromosome.\n",
    "\n",
    "    Args:\n",
    "        num_qtl_per_chromosome (int): Number of QTLs to select per chromosome.\n",
    "        genome (Genome): Genome object containing the chromosome structure.\n",
    "\n",
    "    Returns:\n",
    "        torch.Tensor: A boolean tensor indicating which loci are QTLs. \n",
    "                      Shape: (number_chromosomes, loci_per_chromosome)\n",
    "    \"\"\"\n",
    "    \n",
    "    assert num_qtl_per_chromosome <= genome.n_loci_per_chromosome, \"Too many QTLs for this trait given your Genome object\"\n",
    "    assert num_qtl_per_chromosome > 0, \"You need at least 1 QTL per chromosome\"\n",
    "    \n",
    "    qtl_indices = []\n",
    "    for _ in range(genome.n_chromosomes):\n",
    "        chromosome_indices = torch.randperm(genome.n_loci_per_chromosome)[:num_qtl_per_chromosome]\n",
    "        chromosome_qtl_flags = torch.zeros(genome.n_loci_per_chromosome, dtype=torch.bool)\n",
    "        chromosome_qtl_flags[chromosome_indices] = True\n",
    "        qtl_indices.append(chromosome_qtl_flags)\n",
    "    \n",
    "    return torch.stack(qtl_indices).to(genome.device)\n",
    "\n",
    "\n",
    "\n",
    "class TraitModule(nn.Module):\n",
    "    \"\"\"\n",
    "    Module for managing and simulating multiple correlated additive traits.\n",
    "    \"\"\"\n",
    "    def __init__(self, genome: Genome,founder_pop, target_means: torch.Tensor, target_vars: torch.Tensor, \n",
    "                 correlation_matrix: torch.Tensor, n_qtl_per_chromosome: int):\n",
    "        \"\"\"\n",
    "        Initializes the TraitModule.\n",
    "\n",
    "        Args:\n",
    "            genome (Genome): The genome object.\n",
    "            target_means (torch.Tensor): Target means for each trait (n_traits).\n",
    "            target_vars (torch.Tensor): Target variances for each trait (n_traits).\n",
    "            correlation_matrix (torch.Tensor): Correlation matrix between traits (n_traits, n_traits).\n",
    "            n_qtl_per_chromosome (int): Number of QTLs per chromosome for each trait.\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "        self.genome = genome\n",
    "        self.founder_pop = founder_pop\n",
    "        self.n_traits = len(target_means)\n",
    "        self.target_means = target_means.to(genome.device)\n",
    "        self.target_vars = target_vars.to(genome.device)\n",
    "        self.correlation_matrix = correlation_matrix.to(genome.device)\n",
    "        self.n_qtl_per_chromosome = n_qtl_per_chromosome\n",
    "        \n",
    "        self.qtl_loci = select_qtl_loci(n_qtl_per_chromosome, genome)\n",
    "        self.effects = self._initialize_correlated_effects()\n",
    "        self.intercepts = self._calculate_intercepts()\n",
    "        \n",
    "\n",
    "    def _initialize_correlated_effects(self) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Samples and scales correlated additive effects for all traits.\n",
    "\n",
    "        Returns:\n",
    "            torch.Tensor: Correlated effects (n_chromosomes, n_loci_per_chromosome, n_traits).\n",
    "        \"\"\"\n",
    "        n_chr, n_loci = self.genome.genetic_map.shape\n",
    "        \n",
    "        L = torch.linalg.cholesky(self.correlation_matrix)\n",
    "\n",
    "        uncorrelated_effects = torch.randn(n_chr, n_loci, self.n_traits, device=self.genome.device)\n",
    "        uncorrelated_effects = uncorrelated_effects.reshape(n_chr * n_loci, self.n_traits)\n",
    "\n",
    "        correlated_effects = torch.matmul(L, uncorrelated_effects.T).T\n",
    "        return correlated_effects.reshape(n_chr, n_loci, self.n_traits)\n",
    "\n",
    "    def _calculate_intercepts(self) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Calculates intercepts for each trait to achieve the target means.\n",
    "        \n",
    "        Note:\n",
    "            This calculation depends on an initial population to get unscaled means and variances.\n",
    "            You should call this method after creating a founder population.\n",
    "\n",
    "        Returns:\n",
    "            torch.Tensor: Trait intercepts (n_traits).\n",
    "        \"\"\"\n",
    "        # Example: Calculate intercepts based on a founder population\n",
    "        dosages = self.founder_pop.get_dosages()\n",
    "        unscaled_bvs = self.calculate_breeding_values(dosages, scale_effects=False)\n",
    "        unscaled_var = unscaled_bvs.var(dim=0, unbiased=False)\n",
    "        unscaled_mean = unscaled_bvs.mean(dim=0)\n",
    "        \n",
    "        scaling_factors = torch.sqrt(self.target_vars / unscaled_var)\n",
    "#         import pdb; pdb.set_trace()\n",
    "        self.effects *= scaling_factors.view(1, 1, 3)  # Scale the effects\n",
    "        return self.target_means - (unscaled_mean * scaling_factors)\n",
    "\n",
    "    def calculate_breeding_values(self, dosages: torch.Tensor, scale_effects: bool = True) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Calculates breeding values for all traits given allele dosages.\n",
    "\n",
    "        Args:\n",
    "            dosages (torch.Tensor): Allele dosages (population_size, n_chromosomes, n_loci_per_chromosome).\n",
    "            scale_effects (bool): Whether to scale effects to target variances. Defaults to True.\n",
    "\n",
    "        Returns:\n",
    "            torch.Tensor: Breeding values for all traits (population_size, n_traits).\n",
    "        \"\"\"\n",
    "        if scale_effects:\n",
    "            return torch.einsum('ijk,jkl->il', dosages.float(), self.effects) + self.intercepts \n",
    "        else:\n",
    "            return torch.einsum('ijk,jkl->il', dosages.float(), self.effects)\n",
    "    \n",
    "    def forward(self, dosages: torch.Tensor, h2: Optional[Union[float, torch.Tensor]] = None, \n",
    "                varE: Optional[Union[float, torch.Tensor]] = None) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Calculates breeding values and adds environmental noise.\n",
    "\n",
    "        Args:\n",
    "            dosages (torch.Tensor): Allele dosages (pop_size, n_chr, n_loci).\n",
    "            h2 (Optional[Union[float, torch.Tensor]]): Heritability (single value or per trait). \n",
    "            varE (Optional[Union[float, torch.Tensor]]): Environmental variance (single value or per trait).\n",
    "\n",
    "        Returns:\n",
    "            torch.Tensor: Phenotypes (pop_size, n_traits).\n",
    "        \"\"\"\n",
    "        breeding_values = self.calculate_breeding_values(dosages.float())\n",
    "        \n",
    "        # Add environmental noise\n",
    "        if varE is not None:\n",
    "            if isinstance(varE, float):\n",
    "                varE = torch.tensor([varE] * self.n_traits, device=self.genome.device)\n",
    "            env_noise = torch.randn_like(breeding_values) * torch.sqrt(varE)\n",
    "            return breeding_values + env_noise \n",
    "        elif h2 is not None:\n",
    "            if isinstance(h2, float):\n",
    "                h2 = torch.tensor([h2] * self.n_traits, device=self.genome.device)\n",
    "            varG = breeding_values.var(dim=0, unbiased=False)\n",
    "            varE = varG * (1 - h2) / h2\n",
    "            env_noise = torch.randn_like(breeding_values) * torch.sqrt(varE)\n",
    "            return breeding_values + env_noise\n",
    "        else:\n",
    "            return breeding_values  # No noise added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a757cf4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created genetic map\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "torch.Size([333, 3])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ploidy = 2\n",
    "n_chr = 10\n",
    "n_loci = 100\n",
    "n_Ind = 333\n",
    "g = Genome(ploidy, n_chr, n_loci)\n",
    "population = Population()\n",
    "population.create_random_founder_population(g, n_founders=n_Ind)\n",
    "init_pop = population.get_dosages().float()  # gets allele dosage for calculating trait values\n",
    "\n",
    "# multi_traits\n",
    "target_means = torch.tensor([0, 5, 20])\n",
    "target_vars = torch.tensor([1, 1, 0.5])  # Note: I'm assuming you want a variance of 1 for the second trait\n",
    "correlation_matrix = [\n",
    "        [1.0, 0.2, 0.58],\n",
    "        [0.2, 1.0, -0.37],\n",
    "        [0.58, -0.37, 1.0],\n",
    "    ]\n",
    "correlation_matrix = torch.tensor(correlation_matrix)\n",
    "\n",
    "ta = TraitModule(g, population, target_means, target_vars, correlation_matrix,100)\n",
    "ta(population.get_dosages()).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f836f5b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
