{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3444796d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp chewc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f34d42d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from fastcore.basics import patch\n",
    "import uuid\n",
    "\n",
    "class ChewC:\n",
    "    def __init___(self):\n",
    "        self.trait = None\n",
    "        self.genome = None\n",
    "        self.population = None\n",
    "\n",
    "        \n",
    "class Genome:\n",
    "    def __init__(self, n_chr, n_loci):\n",
    "        self.ploidy = 2\n",
    "        self.n_chr = n_chr\n",
    "        self.n_loci = n_loci\n",
    "        self.shape = (self.ploidy, self.n_chr, self.n_loci)\n",
    "        \n",
    "class Trait:\n",
    "    def __init__(self, genome, founder_population, target_mean, target_variance):\n",
    "        #store attributes\n",
    "        self.target_mean = target_mean\n",
    "        self.target_variance = target_variance\n",
    "        \n",
    "        #sample initial random effects\n",
    "        random_effects = np.random.randn(g.n_chr, g.n_loci)\n",
    "        #calculate the founder_population mean and var given these effects\n",
    "        founder_scores = np.array([random_effects @ x.haplotype for x in population.individuals])\n",
    "        founder_mean, founder_var = founder_scores.mean(), founder_scores.var()\n",
    "        #scale the random effects to match our target variance\n",
    "        scaling_factors = np.sqrt(self.target_variance / founder_var)\n",
    "        random_effects *= scaling_factors\n",
    "        self.effects = random_effects\n",
    "        \n",
    "        self.intercept = founder_mean - target_mean\n",
    "        \n",
    "    def __matmul__(self,other):\n",
    "        if isinstance(other,Individual):\n",
    "            return self.effects * other.haplotype + self.intercept\n",
    "        elif isinstance(other, Population):\n",
    "            return np.sum(np.array([self @ ind + self.intercept for ind in other.individuals]), axis=(1,2))\n",
    "\n",
    "        \n",
    "class Population:\n",
    "    def __init__(self, genome, size=None):\n",
    "        self.genome = genome\n",
    "        self.ploidy = 2\n",
    "        if size:\n",
    "            self.size = size\n",
    "            self.individuals = self._create_initial_population()\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "    def _create_initial_population(self):\n",
    "        \"\"\"Create an initial population of founder individuals.\"\"\"\n",
    "        return [Individual(self.genome) for _ in range(self.size)]\n",
    "    \n",
    "    def get_haplo(self):\n",
    "        return np.array([x.haplotype for x in self.individuals])\n",
    "    \n",
    "    def get_pheno(self):\n",
    "        return np.array([x.fitness for x in self.individuals])\n",
    "    \n",
    "    def __getitem__(self, index):\n",
    "        # Convert numpy arrays to lists for indexing\n",
    "        if isinstance(index, np.ndarray):\n",
    "            index = index.tolist()\n",
    "        # Handle list of indices\n",
    "        if isinstance(index, list):\n",
    "            return [self.individuals[i] for i in index]\n",
    "        # Handle single index\n",
    "        return self.individuals[index]\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return f'Population of size: {self.size}'\n",
    "    \n",
    "\n",
    "class Individual:\n",
    "    def __init__(self, genome,haplotype=None, mother = None, father = None, descendants = 0, source='founder', chewc=None):\n",
    "        self.genome = genome  # must be Genome class\n",
    "        self.haplotype = None\n",
    "        self.source = source  # 'founder', 'cross', 'self', 'dh'\n",
    "        self.descendents = []\n",
    "        self.fitness = 0\n",
    "        self.mother, self.father = mother, father\n",
    "\n",
    "        # Logic for setting haplotype,mother,father\n",
    "        if self.source == 'founder':\n",
    "            self.haplotype= self._generate_random_haplotype()\n",
    "        else:\n",
    "            self.haplotype = haplotype\n",
    "            \n",
    "        self.make_id()\n",
    "            \n",
    "    def __repr__(self):\n",
    "        return f'Individual with haplotype shape: {self.haplotype.shape}'\n",
    "    \n",
    "    def make_id(self):self.id = uuid.uuid4().hex\n",
    "\n",
    "\n",
    "    def _generate_random_haplotype(self):\n",
    "        \"\"\"Generate a random haplotype for the individual.\"\"\"\n",
    "        return np.random.choice([0, 1], size=(self.genome.ploidy, self.genome.n_chr, self.genome.n_loci))\n",
    "         \n",
    "\n",
    "# to do move this in another module\n",
    "@patch\n",
    "def gamete(self:Individual):\n",
    "    haplotypes = self.haplotype\n",
    "    def shuffle_chr(chromosome_pair):\n",
    "        \"\"\"\n",
    "        Perform crossover on a pair of chromosomes.\n",
    "\n",
    "        Parameters:\n",
    "        chromosome_pair (list): A list of two equal length lists representing chromosomes.\n",
    "\n",
    "        Returns:\n",
    "        list: A new chromosome formed by shuffling the given chromosome pair.\n",
    "        \"\"\"\n",
    "        # Ensure the chromosome pair contains two chromosomes of equal length\n",
    "        assert len(chromosome_pair) == 2\n",
    "        assert len(chromosome_pair[0]) == len(chromosome_pair[1])\n",
    "\n",
    "        # Number of crossover points, sampled from a Poisson distribution with Î»=1.3\n",
    "        n_crossover = np.random.poisson(1.3)\n",
    "\n",
    "        # Determine crossover locations, sampled without replacement from chromosome length\n",
    "        chromosome_length = len(chromosome_pair[0])\n",
    "        crossover_locs = np.sort(np.random.choice(chromosome_length, n_crossover, replace=False))\n",
    "\n",
    "        # Initialize the new chromosome and set the current chromosome to the first one\n",
    "        new_chromosome = []\n",
    "        current_chr = 0\n",
    "\n",
    "        # Perform crossover by alternating segments between the two chromosomes\n",
    "        last_loc = 0\n",
    "        for loc in crossover_locs:\n",
    "            new_chromosome.extend(chromosome_pair[current_chr][last_loc:loc])\n",
    "            current_chr = 1 - current_chr  # Switch to the other chromosome\n",
    "            last_loc = loc\n",
    "\n",
    "        # Append the remaining segment\n",
    "        new_chromosome.extend(chromosome_pair[current_chr][last_loc:])\n",
    "        return np.array(new_chromosome)\n",
    "\n",
    "    # Initialize an empty array to store the shuffled chromosomes\n",
    "    shuffled_haplotypes = np.zeros_like(haplotypes)\n",
    "    ploidy, n_chr, n_loci = chewc.genome.shape\n",
    "    # Iterate over each chromosome and apply the shuffle_chr function\n",
    "    for i in range(haplotypes.shape[1]):  # Iterate over the chromosomes\n",
    "        chromosome_pair = haplotypes[:, i, :]  # Extract the chromosome pair (2, 77)\n",
    "        shuffled_chromosome = shuffle_chr(chromosome_pair)  # Shuffle the chromosome pair    \n",
    "        shuffled_haplotypes[:, i, :] = np.array(shuffled_chromosome).reshape(1, n_loci)  # Store the shuffled chromosome\n",
    "    return shuffled_haplotypes[0,:,:]\n",
    "\n",
    "class Trait:\n",
    "    def __init__(self, genome, founder_population, target_mean, target_variance):\n",
    "        #store attributes\n",
    "        self.target_mean = target_mean\n",
    "        self.target_variance = target_variance\n",
    "        \n",
    "        #sample initial random effects\n",
    "        initial_effects = np.random.randn(g.n_chr * g.n_loci)\n",
    "        initial_effects -= initial_effects.mean()\n",
    "        #calculate the founder_population mean and var given these effects\n",
    "        pop_scores = np.sum(founder_population.get_haplo(),axis=1).reshape(founder_population.size, g.n_chr*g.n_loci)\n",
    "        founder_scores = np.sum(initial_effects*pop_scores,axis=1)\n",
    "        founder_mean, founder_var = founder_scores.mean(), founder_scores.var()\n",
    "        scaling_factors = np.sqrt(self.target_variance / founder_var)\n",
    "        initial_effects *= scaling_factors\n",
    "        scaled_effects = initial_effects # ADDS A FIXED EFFECT SOMEHOW????\n",
    "        self.effects = scaled_effects\n",
    "        self.intercept = target_mean - founder_mean\n",
    "\n",
    "@patch\n",
    "def x(self:Individual, partner):\n",
    "    if isinstance(partner,Individual):\n",
    "        source = 'cross'\n",
    "        progeny_haplo = np.vstack((self.gamete(), partner.gamete()))\n",
    "        \n",
    "        progeny = Individual(self.genome, progeny_haplo, self.id, partner.id,source=source, chewc = chewc)\n",
    "        return progeny\n",
    "    \n",
    "@patch\n",
    "def __matmul__(self:Trait,other):\n",
    "    if isinstance(other,Individual):\n",
    "#             print(f' intercept {self.intercept}')\n",
    "        breeding_value = self.effects * np.sum(other.haplotype,axis=0).flatten()\n",
    "        return breeding_value\n",
    "    else:\n",
    "        print('ffff')\n",
    "@patch\n",
    "def phenotype(self:Individual, h2, environmental_variance):\n",
    "    \"\"\"\n",
    "    Calculate the phenotype for the individual.\n",
    "\n",
    "    Args:\n",
    "        h2 (float): Heritability of the trait.\n",
    "        environmental_variance (float): The calculated environmental variance for the population.\n",
    "    \"\"\"\n",
    "    breeding_value = chewc.trait @ self\n",
    "    phenotype_value = breeding_value + np.random.normal(0, np.sqrt(environmental_variance))\n",
    "    self.fitness = np.sum(phenotype_value)\n",
    "    \n",
    "@patch\n",
    "def trial(self:Population, h2):\n",
    "    \"\"\"\n",
    "    Simulate phenotypes for all individuals in the population.\n",
    "\n",
    "    Args:\n",
    "        h2 (float): Heritability of the trait.\n",
    "    \"\"\"\n",
    "    # 1. Calculate environmental variance at the population level\n",
    "    population_genetic_variance = np.var(chewc.population.get_pheno(), ddof=1)\n",
    "    environmental_variance = (1 - h2) / h2 * population_genetic_variance\n",
    "\n",
    "    # 2. Phenotype each individual, passing the environmental variance\n",
    "    for individual in self.individuals:\n",
    "        individual.phenotype(h2, environmental_variance) \n",
    "\n",
    "    def __getitem__(self, index):\n",
    "        return self.individuals[index]\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'Population of size: {self.size}'\n",
    "\n",
    "@patch\n",
    "def truncation(self:Population, top_percent):\n",
    "    \"\"\"\n",
    "    Returns the indexes of phenotypes in the top percent.\n",
    "\n",
    "    Args:\n",
    "        top_percent (float): The top percentage of individuals to select (0 < top_percent <= 100).\n",
    "    \"\"\"\n",
    "    # Ensure the top_percent is within the valid range\n",
    "    if top_percent <= 0 or top_percent > 100:\n",
    "        raise ValueError(\"top_percent must be between 0 and 100\")\n",
    "\n",
    "    # Calculate the number of individuals to select\n",
    "    num_to_select = int(population.size * top_percent / 100)\n",
    "\n",
    "    # Get the phenotypes of all individuals\n",
    "    phenotypes = self.get_pheno()\n",
    "    assert phenotypes.sum() != 0, 'no phenotypes present'\n",
    "#     print(phenotypes)\n",
    "\n",
    "    # Get the indices that would sort the array in descending order\n",
    "    sorted_indices = np.argsort(phenotypes)[::-1]\n",
    "\n",
    "    # Select the top individuals based on the sorted indices\n",
    "    top_indices = sorted_indices[:num_to_select]\n",
    "\n",
    "    return self[top_indices]\n",
    "\n",
    "\n",
    "#Define the Simulation Parameters\n",
    "g = Genome(3, 1000)\n",
    "population = Population(g, size=999)\n",
    "trait = Trait(g, population,5,1000)\n",
    "\n",
    "#Plug them into ChewC\n",
    "chewc = ChewC()\n",
    "chewc.trait = trait\n",
    "chewc.population = population\n",
    "chewc.genome = g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff8aa8cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "chewc.population.trial(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1b32e029",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26.08365894945355"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chewc.population[0].fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fc80ef41",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([  8.,  45., 113., 223., 267., 206.,  93.,  37.,   6.,   1.]),\n",
       " array([-93.5297071 , -72.15439126, -50.77907542, -29.40375958,\n",
       "         -8.02844374,  13.3468721 ,  34.72218794,  56.09750378,\n",
       "         77.47281962,  98.84813546, 120.2234513 ]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAGdCAYAAAA44ojeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy80BEi2AAAACXBIWXMAAA9hAAAPYQGoP6dpAAAghklEQVR4nO3de3BU5cHH8V8SyEKA3Rgg2UQDBC9c5Cpo3IoMlpQQUiol7QBNER0K1SZ2IBQxVrnYaig64ugg1JmW2Kl4YabigBobw62UJWqEIhczQMFgYROFJgtUciHP+0eH83YFxUDCPkm+n5kzZs959uQ5HGO+nj27RBhjjAAAACwSGe4JAAAAfBWBAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6HcI9gcvR2NioY8eOqVu3boqIiAj3dAAAwLdgjNGpU6eUlJSkyMhvvkbSKgPl2LFjSk5ODvc0AADAZTh69Kiuu+66bxzTKgOlW7dukv57gG63O8yzAQAA30YwGFRycrLze/ybtMpAOf+yjtvtJlAAAGhlvs3tGdwkCwAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA63QI9wQAtKw+D78V7ik02ZGlmeGeAoAw4woKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKzTpEApKCjQrbfeqm7duik+Pl6TJk1SeXl5yJgxY8YoIiIiZLn//vtDxlRUVCgzM1MxMTGKj4/X/Pnz1dDQcOVHAwAA2oQOTRm8ZcsW5eTk6NZbb1VDQ4MeeeQRjRs3Tvv27VOXLl2ccbNmzdLjjz/uPI6JiXG+PnfunDIzM+X1erV9+3YdP35c99xzjzp27Kgnn3yyGQ4JAAC0dk0KlKKiopDHhYWFio+PV1lZmUaPHu2sj4mJkdfrveg+/vrXv2rfvn167733lJCQoGHDhuk3v/mNFixYoMWLFys6OvoyDgMAALQlV3QPSk1NjSQpLi4uZP3LL7+sHj16aNCgQcrPz9d//vMfZ5vf79fgwYOVkJDgrEtPT1cwGNTevXsv+n1qa2sVDAZDFgAA0HY16QrK/2psbNScOXN0xx13aNCgQc76n/zkJ+rdu7eSkpK0e/duLViwQOXl5frLX/4iSQoEAiFxIsl5HAgELvq9CgoKtGTJksudKgAAaGUuO1BycnK0Z88ebdu2LWT97Nmzna8HDx6sxMREjR07VocOHdL1119/Wd8rPz9feXl5zuNgMKjk5OTLmzgAALDeZb3Ek5ubqw0bNmjTpk267rrrvnFsamqqJOngwYOSJK/Xq8rKypAx5x9/3X0rLpdLbrc7ZAEAAG1XkwLFGKPc3Fy98cYb2rhxo1JSUi75nF27dkmSEhMTJUk+n08ff/yxqqqqnDHFxcVyu90aOHBgU6YDAADaqCa9xJOTk6M1a9bozTffVLdu3Zx7Rjwejzp37qxDhw5pzZo1mjBhgrp3767du3dr7ty5Gj16tIYMGSJJGjdunAYOHKjp06dr2bJlCgQCevTRR5WTkyOXy9X8RwgAAFqdJl1BWblypWpqajRmzBglJiY6y2uvvSZJio6O1nvvvadx48apf//+mjdvnrKysrR+/XpnH1FRUdqwYYOioqLk8/n005/+VPfcc0/I56YAAID2rUlXUIwx37g9OTlZW7ZsueR+evfurbfffrsp3xoAALQj/F08AADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrdAj3BIDWpM/Db4V7CgDQLnAFBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANbpEO4JAMBX9Xn4rXBPocmOLM0M9xSANoUrKAAAwDoECgAAsE6TAqWgoEC33nqrunXrpvj4eE2aNEnl5eUhY86ePaucnBx1795dXbt2VVZWliorK0PGVFRUKDMzUzExMYqPj9f8+fPV0NBw5UcDAADahCYFypYtW5STk6MdO3aouLhY9fX1GjdunM6cOeOMmTt3rtavX6+1a9dqy5YtOnbsmCZPnuxsP3funDIzM1VXV6ft27frpZdeUmFhoRYuXNh8RwUAAFq1CGOMudwnf/7554qPj9eWLVs0evRo1dTUqGfPnlqzZo1+9KMfSZI++eQTDRgwQH6/X7fffrveeecdff/739exY8eUkJAgSVq1apUWLFigzz//XNHR0Zf8vsFgUB6PRzU1NXK73Zc7faDJWuPNm7g6uEkWuLSm/P6+ontQampqJElxcXGSpLKyMtXX1ystLc0Z079/f/Xq1Ut+v1+S5Pf7NXjwYCdOJCk9PV3BYFB79+69kukAAIA24rLfZtzY2Kg5c+bojjvu0KBBgyRJgUBA0dHRio2NDRmbkJCgQCDgjPnfODm//fy2i6mtrVVtba3zOBgMXu60AQBAK3DZV1BycnK0Z88evfrqq805n4sqKCiQx+NxluTk5Bb/ngAAIHwuK1Byc3O1YcMGbdq0Sdddd52z3uv1qq6uTtXV1SHjKysr5fV6nTFffVfP+cfnx3xVfn6+ampqnOXo0aOXM20AANBKNClQjDHKzc3VG2+8oY0bNyolJSVk+4gRI9SxY0eVlJQ468rLy1VRUSGfzydJ8vl8+vjjj1VVVeWMKS4ultvt1sCBAy/6fV0ul9xud8gCAADaribdg5KTk6M1a9bozTffVLdu3Zx7Rjwejzp37iyPx6OZM2cqLy9PcXFxcrvdevDBB+Xz+XT77bdLksaNG6eBAwdq+vTpWrZsmQKBgB599FHl5OTI5XI1/xECAIBWp0mBsnLlSknSmDFjQtavXr1a9957ryRp+fLlioyMVFZWlmpra5Wenq4XXnjBGRsVFaUNGzbogQcekM/nU5cuXTRjxgw9/vjjV3YkAACgzbiiz0EJFz4HBeHC56Dg6/A5KMClXbXPQQEAAGgJBAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArNPkQNm6dasmTpyopKQkRUREaN26dSHb7733XkVERIQs48ePDxlz8uRJZWdny+12KzY2VjNnztTp06ev6EAAAEDb0eRAOXPmjIYOHaoVK1Z87Zjx48fr+PHjzvLKK6+EbM/OztbevXtVXFysDRs2aOvWrZo9e3bTZw8AANqkDk19QkZGhjIyMr5xjMvlktfrvei2/fv3q6ioSB988IFGjhwpSXr++ec1YcIEPf3000pKSmrqlAAAQBvTIvegbN68WfHx8erXr58eeOABnThxwtnm9/sVGxvrxIkkpaWlKTIyUqWlpS0xHQAA0Mo0+QrKpYwfP16TJ09WSkqKDh06pEceeUQZGRny+/2KiopSIBBQfHx86CQ6dFBcXJwCgcBF91lbW6va2lrncTAYbO5pAwAAizR7oEydOtX5evDgwRoyZIiuv/56bd68WWPHjr2sfRYUFGjJkiXNNUUAAGC5Fn+bcd++fdWjRw8dPHhQkuT1elVVVRUypqGhQSdPnvza+1by8/NVU1PjLEePHm3paQMAgDBq8UD57LPPdOLECSUmJkqSfD6fqqurVVZW5ozZuHGjGhsblZqaetF9uFwuud3ukAUAALRdTX6J5/Tp087VEEk6fPiwdu3apbi4OMXFxWnJkiXKysqS1+vVoUOH9NBDD+mGG25Qenq6JGnAgAEaP368Zs2apVWrVqm+vl65ubmaOnUq7+ABAACSLuMKyocffqjhw4dr+PDhkqS8vDwNHz5cCxcuVFRUlHbv3q0f/OAHuummmzRz5kyNGDFCf/vb3+RyuZx9vPzyy+rfv7/Gjh2rCRMmaNSoUXrxxReb76gAAECr1uQrKGPGjJEx5mu3v/vuu5fcR1xcnNasWdPUbw0AANoJ/i4eAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFinQ7gngParz8NvhXsKAABLcQUFAABYh0ABAADWIVAAAIB1CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgnSYHytatWzVx4kQlJSUpIiJC69atC9lujNHChQuVmJiozp07Ky0tTQcOHAgZc/LkSWVnZ8vtdis2NlYzZ87U6dOnr+hAAABA29HkQDlz5oyGDh2qFStWXHT7smXL9Nxzz2nVqlUqLS1Vly5dlJ6errNnzzpjsrOztXfvXhUXF2vDhg3aunWrZs+efflHAQAA2pQOTX1CRkaGMjIyLrrNGKNnn31Wjz76qO6++25J0p/+9CclJCRo3bp1mjp1qvbv36+ioiJ98MEHGjlypCTp+eef14QJE/T0008rKSnpCg4HAAC0Bc16D8rhw4cVCASUlpbmrPN4PEpNTZXf75ck+f1+xcbGOnEiSWlpaYqMjFRpaelF91tbW6tgMBiyAACAtqtZAyUQCEiSEhISQtYnJCQ42wKBgOLj40O2d+jQQXFxcc6YryooKJDH43GW5OTk5pw2AACwTKt4F09+fr5qamqc5ejRo+GeEgAAaEHNGiher1eSVFlZGbK+srLS2eb1elVVVRWyvaGhQSdPnnTGfJXL5ZLb7Q5ZAABA29WsgZKSkiKv16uSkhJnXTAYVGlpqXw+nyTJ5/OpurpaZWVlzpiNGzeqsbFRqampzTkdAADQSjX5XTynT5/WwYMHnceHDx/Wrl27FBcXp169emnOnDn67W9/qxtvvFEpKSl67LHHlJSUpEmTJkmSBgwYoPHjx2vWrFlatWqV6uvrlZubq6lTp/IOHgAAIOkyAuXDDz/UXXfd5TzOy8uTJM2YMUOFhYV66KGHdObMGc2ePVvV1dUaNWqUioqK1KlTJ+c5L7/8snJzczV27FhFRkYqKytLzz33XDMcDgAAaAsijDEm3JNoqmAwKI/Ho5qaGu5HacX6PPxWuKcANJsjSzPDPQXAek35/d0q3sUDAADaFwIFAABYh0ABAADWafJNsgCAC7XGe6q4bwY24woKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrNHugLF68WBERESFL//79ne1nz55VTk6Ounfvrq5duyorK0uVlZXNPQ0AANCKtcgVlJtvvlnHjx93lm3btjnb5s6dq/Xr12vt2rXasmWLjh07psmTJ7fENAAAQCvVoUV22qGDvF7vBetramr0hz/8QWvWrNF3v/tdSdLq1as1YMAA7dixQ7fffntLTAcAALQyLXIF5cCBA0pKSlLfvn2VnZ2tiooKSVJZWZnq6+uVlpbmjO3fv7969eolv9//tfurra1VMBgMWQAAQNvV7IGSmpqqwsJCFRUVaeXKlTp8+LDuvPNOnTp1SoFAQNHR0YqNjQ15TkJCggKBwNfus6CgQB6Px1mSk5Obe9oAAMAizf4ST0ZGhvP1kCFDlJqaqt69e+v1119X586dL2uf+fn5ysvLcx4Hg0EiBQCANqzF32YcGxurm266SQcPHpTX61VdXZ2qq6tDxlRWVl70npXzXC6X3G53yAIAANquFg+U06dP69ChQ0pMTNSIESPUsWNHlZSUONvLy8tVUVEhn8/X0lMBAACtRLO/xPOrX/1KEydOVO/evXXs2DEtWrRIUVFRmjZtmjwej2bOnKm8vDzFxcXJ7XbrwQcflM/n4x08AADA0eyB8tlnn2natGk6ceKEevbsqVGjRmnHjh3q2bOnJGn58uWKjIxUVlaWamtrlZ6erhdeeKG5pwEAAFqxCGOMCfckmioYDMrj8aimpob7UVqxPg+/Fe4pAO3akaWZ4Z4C2pmm/P5ukQ9qw9XHL3sAQFvCXxYIAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArMMnyQJAO9UaP4Gaj+dvP7iCAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA6xAoAADAOgQKAACwDoECAACsQ6AAAADrECgAAMA6BAoAALAOgQIAAKxDoAAAAOsQKAAAwDoECgAAsA6BAgAArEOgAAAA63QI9wRs1Ofht8I9BQAA2jWuoAAAAOtwBQUA0Gq0xivcR5ZmhnsKrRJXUAAAgHUIFAAAYB0CBQAAWCesgbJixQr16dNHnTp1Umpqqt5///1wTgcAAFgibIHy2muvKS8vT4sWLdJHH32koUOHKj09XVVVVeGaEgAAsETYAuWZZ57RrFmzdN9992ngwIFatWqVYmJi9Mc//jFcUwIAAJYIy9uM6+rqVFZWpvz8fGddZGSk0tLS5Pf7LxhfW1ur2tpa53FNTY0kKRgMtsj8Gmv/0yL7BQC0P73mrg33FC7LniXpzb7P87+3jTGXHBuWQPniiy907tw5JSQkhKxPSEjQJ598csH4goICLVmy5IL1ycnJLTZHAADaM8+zLbfvU6dOyePxfOOYVvFBbfn5+crLy3MeNzY26uTJk+revbsiIiIu+pxgMKjk5GQdPXpUbrf7ak0VTcA5ah04T/bjHLUOnKf/Xjk5deqUkpKSLjk2LIHSo0cPRUVFqbKyMmR9ZWWlvF7vBeNdLpdcLlfIutjY2G/1vdxud7v9F6G14By1Dpwn+3GOWof2fp4udeXkvLDcJBsdHa0RI0aopKTEWdfY2KiSkhL5fL5wTAkAAFgkbC/x5OXlacaMGRo5cqRuu+02Pfvsszpz5ozuu+++cE0JAABYImyBMmXKFH3++edauHChAoGAhg0bpqKiogtunL1cLpdLixYtuuClIdiDc9Q6cJ7sxzlqHThPTRNhvs17fQAAAK4i/i4eAABgHQIFAABYh0ABAADWIVAAAIB1Wn2gPPHEE/rOd76jmJiYr/3wtoqKCmVmZiomJkbx8fGaP3++GhoaQsZs3rxZt9xyi1wul2644QYVFha2/OTbsT59+igiIiJkWbp0aciY3bt3684771SnTp2UnJysZcuWhWm27deKFSvUp08fderUSampqXr//ffDPaV2bfHixRf83PTv39/ZfvbsWeXk5Kh79+7q2rWrsrKyLvhATDSvrVu3auLEiUpKSlJERITWrVsXst0Yo4ULFyoxMVGdO3dWWlqaDhw4EDLm5MmTys7OltvtVmxsrGbOnKnTp09fxaOwU6sPlLq6Ov34xz/WAw88cNHt586dU2Zmpurq6rR9+3a99NJLKiws1MKFC50xhw8fVmZmpu666y7t2rVLc+bM0c9+9jO9++67V+sw2qXHH39cx48fd5YHH3zQ2RYMBjVu3Dj17t1bZWVleuqpp7R48WK9+OKLYZxx+/Laa68pLy9PixYt0kcffaShQ4cqPT1dVVVV4Z5au3bzzTeH/Nxs27bN2TZ37lytX79ea9eu1ZYtW3Ts2DFNnjw5jLNt+86cOaOhQ4dqxYoVF92+bNkyPffcc1q1apVKS0vVpUsXpaen6+zZs86Y7Oxs7d27V8XFxdqwYYO2bt2q2bNnX61DsJdpI1avXm08Hs8F699++20TGRlpAoGAs27lypXG7Xab2tpaY4wxDz30kLn55ptDnjdlyhSTnp7eonNuz3r37m2WL1/+tdtfeOEFc8011zjnyBhjFixYYPr163cVZgdjjLnttttMTk6O8/jcuXMmKSnJFBQUhHFW7duiRYvM0KFDL7qturradOzY0axdu9ZZt3//fiPJ+P3+qzTD9k2SeeONN5zHjY2Nxuv1mqeeespZV11dbVwul3nllVeMMcbs27fPSDIffPCBM+add94xERER5l//+tdVm7uNWv0VlEvx+/0aPHhwyAfApaenKxgMau/evc6YtLS0kOelp6fL7/df1bm2N0uXLlX37t01fPhwPfXUUyEvu/n9fo0ePVrR0dHOuvT0dJWXl+vf//53OKbbrtTV1amsrCzk5yIyMlJpaWn8XITZgQMHlJSUpL59+yo7O1sVFRWSpLKyMtXX14ecs/79+6tXr16cszA5fPiwAoFAyDnxeDxKTU11zonf71dsbKxGjhzpjElLS1NkZKRKS0uv+pxt0ir+NuMrEQgELvh02vOPA4HAN44JBoP68ssv1blz56sz2Xbkl7/8pW655RbFxcVp+/btys/P1/Hjx/XMM89I+u85SUlJCXnO/563a6655qrPuT354osvdO7cuYv+XHzyySdhmhVSU1NVWFiofv366fjx41qyZInuvPNO7dmzR4FAQNHR0Rfci5eQkOD8tw5X1/k/94v9HP3v75/4+PiQ7R06dFBcXFy7P29WBsrDDz+s3/3ud984Zv/+/SE3hyH8mnLe8vLynHVDhgxRdHS0fv7zn6ugoICPgQa+RkZGhvP1kCFDlJqaqt69e+v111/nf6TQ5lgZKPPmzdO99977jWP69u37rfbl9XoveOfB+bvavV6v88+v3uleWVkpt9vND30TXMl5S01NVUNDg44cOaJ+/fp97TmR/v+8oeX06NFDUVFRFz0H/PnbIzY2VjfddJMOHjyo733ve6qrq1N1dXXIVRTOWfic/3OvrKxUYmKis76yslLDhg1zxnz1xvOGhgadPHmy3Z83KwOlZ8+e6tmzZ7Psy+fz6YknnlBVVZVzGa24uFhut1sDBw50xrz99tshzysuLpbP52uWObQXV3Ledu3apcjISOcc+Xw+/frXv1Z9fb06duwo6b/npF+/fry8cxVER0drxIgRKikp0aRJkyRJjY2NKikpUW5ubngnB8fp06d16NAhTZ8+XSNGjFDHjh1VUlKirKwsSVJ5ebkqKir4b1mYpKSkyOv1qqSkxAmSYDCo0tJS552nPp9P1dXVKisr04gRIyRJGzduVGNjo1JTU8M1dTuE+y7dK/Xpp5+anTt3miVLlpiuXbuanTt3mp07d5pTp04ZY4xpaGgwgwYNMuPGjTO7du0yRUVFpmfPniY/P9/Zxz//+U8TExNj5s+fb/bv329WrFhhoqKiTFFRUbgOq03bvn27Wb58udm1a5c5dOiQ+fOf/2x69uxp7rnnHmdMdXW1SUhIMNOnTzd79uwxr776qomJiTG///3vwzjz9uXVV181LpfLFBYWmn379pnZs2eb2NjYkHfE4eqaN2+e2bx5szl8+LD5+9//btLS0kyPHj1MVVWVMcaY+++/3/Tq1cts3LjRfPjhh8bn8xmfzxfmWbdtp06dcn7vSDLPPPOM2blzp/n000+NMcYsXbrUxMbGmjfffNPs3r3b3H333SYlJcV8+eWXzj7Gjx9vhg8fbkpLS822bdvMjTfeaKZNmxauQ7JGqw+UGTNmGEkXLJs2bXLGHDlyxGRkZJjOnTubHj16mHnz5pn6+vqQ/WzatMkMGzbMREdHm759+5rVq1df3QNpR8rKykxqaqrxeDymU6dOZsCAAebJJ580Z8+eDRn3j3/8w4waNcq4XC5z7bXXmqVLl4Zpxu3X888/b3r16mWio6PNbbfdZnbs2BHuKbVrU6ZMMYmJiSY6Otpce+21ZsqUKebgwYPO9i+//NL84he/MNdcc42JiYkxP/zhD83x48fDOOO2b9OmTRf9HTRjxgxjzH/favzYY4+ZhIQE43K5zNixY015eXnIPk6cOGGmTZtmunbtatxut7nvvvuc/8luzyKMMSZMF28AAAAuqs1/DgoAAGh9CBQAAGAdAgUAAFiHQAEAANYhUAAAgHUIFAAAYB0CBQAAWIdAAQAA1iFQAACAdQgUAABgHQIFAABYh0ABAADW+T81ldnp/YaG+QAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "plt.hist(chewc.population.get_pheno())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d1eb3aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%timeit\n",
    "# chewc.population[0].gamete()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d16d13c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# %%timeit\n",
    "# chewc.population[0].x(chewc.population[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f8d1ffc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "482a15c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def randomize_pairing(input_array):\n",
    "    # Convert input array to numpy array if it isn't already\n",
    "    array = np.array(input_array)\n",
    "    \n",
    "    # Shuffle the array\n",
    "    np.random.shuffle(array)\n",
    "    \n",
    "    # Pair the elements\n",
    "    n = len(array)\n",
    "    if n % 2 != 0:\n",
    "        # If the length of the array is odd, add a None to make it even\n",
    "        array = np.append(array, None)\n",
    "        n += 1\n",
    "    \n",
    "    pairs = [(array[i], array[i + 1]) for i in range(0, n, 2)]\n",
    "    \n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6b4cce1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000)]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chewc.population[4:6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7ee16a41",
   "metadata": {},
   "outputs": [],
   "source": [
    "indexes = chewc.population.truncation(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "688296b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000),\n",
       " Individual with haplotype shape: (2, 3, 1000)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64613b23",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
