{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8ded630",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Population haplotypes is on: cpu\n",
      "Trait effects is on: cpu\n",
      "Trait intercept is on: cpu\n",
      "Population scores is on: cpu\n",
      "Variance of scores: 2.0599539279937744\n",
      "Selected haplotypes is on: cpu\n",
      "Original population shape: torch.Size([1000, 2, 10000])\n",
      "Selected haplotypes shape: torch.Size([200, 2, 10000])\n",
      "Gametes is on: cpu\n",
      "Gametes shape: torch.Size([1000, 10000])\n",
      "New population shape: torch.Size([500, 2, 10000])\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGdCAYAAACyzRGfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy80BEi2AAAACXBIWXMAAA9hAAAPYQGoP6dpAAAcaElEQVR4nO3df6yW9X3/8RcIHJhwDkLhHImAdHUF51D5UTziGrWszJCmRGpt4zJLiC7mSAunTeUsq4zGCGm3ylz5ocahS0rozEJba8QYFrHrwB8YE62RypTAwHOwWzlHaTgQOd8//Pakp1rlhnM+h3PO45HcCee6r3Pdb3OL5+nnXNd9Dero6OgIAEAhg3t7AABgYBEfAEBR4gMAKEp8AABFiQ8AoCjxAQAUJT4AgKLEBwBQ1JDeHuD3nTx5MocOHcqoUaMyaNCg3h4HADgFHR0defvttzNhwoQMHvzhaxtnXXwcOnQoEydO7O0xAIDTcODAgVxwwQUfuk/F8XHw4MHccccdefzxx/Ob3/wmn/jEJ7Jp06bMmjUryXvls3LlyjzwwAM5cuRI5s6dmw0bNuSiiy46peOPGjWqc/jq6upKxwMAekFbW1smTpzY+XP8w1QUH7/+9a8zd+7cXHPNNXn88cczbty4vPbaaznvvPM69/nOd76Te++9Nw8//HCmTJmSb33rW5k/f35eeeWVDB8+/CNf47e/aqmurhYfANDHnMopE4MqubHcihUr8vOf/zw/+9nPPvD5jo6OTJgwIV//+tfzjW98I0nS2tqa2traPPTQQ/nSl770ka/R1taWmpqatLa2ig8A6CMq+fld0dUuP/nJTzJr1qzccMMNGT9+fC6//PI88MADnc+/8cYbaW5uzrx58zq31dTUZM6cOdm5c+cHHrO9vT1tbW1dHgBA/1VRfLz++uud52888cQTue222/LVr341Dz/8cJKkubk5SVJbW9vl+2prazuf+32rV69OTU1N58PJpgDQv1UUHydPnsyMGTNy99135/LLL8+tt96aW265JRs3bjztAZqamtLa2tr5OHDgwGkfCwA4+1UUH+eff34uvvjiLtumTZuW/fv3J0nq6uqSJC0tLV32aWlp6Xzu91VVVXWeXOokUwDo/yqKj7lz52bPnj1dtv3yl7/M5MmTkyRTpkxJXV1dtm/f3vl8W1tbnnnmmdTX13fDuABAX1fRpbbLly/PlVdembvvvjtf/OIX8+yzz+b+++/P/fffn+S9y2uWLVuWu+66KxdddFHnpbYTJkzIwoULe2J+AKCPqSg+Zs+ena1bt6apqSnf/va3M2XKlKxduzY33XRT5z7f/OY3c/To0dx66605cuRIrrrqqmzbtu2UPuMDAOj/KvqcjxJ8zgcA9D099jkfAABnSnwAAEWJDwCgKPEBABQlPgCAoiq61Bbgw1y44rGP3GffmgUFJgHOZlY+AICixAcAUJT4AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAinJvF+jD3EsF6IusfAAARYkPAKAo8QEAFCU+AICixAcAUJT4AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAihIfAEBR4gMAKEp8AABFiQ8AoCjxAQAUJT4AgKLEBwBQlPgAAIoSHwBAUeIDAChqSG8PAPSsC1c89pH77FuzoMAkAO+x8gEAFCU+AICixAcAUJT4AACKEh8AQFHiAwAoSnwAAEVVFB9///d/n0GDBnV5TJ06tfP5Y8eOpaGhIWPHjs3IkSOzaNGitLS0dPvQAEDfVfHKx5/+6Z/mzTff7Hz853/+Z+dzy5cvz6OPPppHHnkkO3bsyKFDh3L99dd368AAQN9W8SecDhkyJHV1de/b3tramgcffDCbN2/OtddemyTZtGlTpk2bll27duWKK64482kBgD6v4pWP1157LRMmTMjHP/7x3HTTTdm/f3+SZPfu3Tlx4kTmzZvXue/UqVMzadKk7Ny58w8er729PW1tbV0eAED/VdHKx5w5c/LQQw/lk5/8ZN58882sWrUqf/7nf56XX345zc3NGTZsWEaPHt3le2pra9Pc3PwHj7l69eqsWrXqtIaH/uxU7skC0BdVFB/XXXdd55+nT5+eOXPmZPLkyfm3f/u3jBgx4rQGaGpqSmNjY+fXbW1tmThx4mkdCwA4+53RpbajR4/On/zJn2Tv3r2pq6vL8ePHc+TIkS77tLS0fOA5Ir9VVVWV6urqLg8AoP86o/h455138t///d85//zzM3PmzAwdOjTbt2/vfH7Pnj3Zv39/6uvrz3hQAKB/qOjXLt/4xjfyuc99LpMnT86hQ4eycuXKnHPOOfnyl7+cmpqaLFmyJI2NjRkzZkyqq6uzdOnS1NfXu9IFAOhUUXz8z//8T7785S/nf//3fzNu3LhcddVV2bVrV8aNG5ckueeeezJ48OAsWrQo7e3tmT9/ftavX98jgwMAfdOgjo6Ojt4e4ne1tbWlpqYmra2tzv9gQCt5tcu+NQu65Th9cWage1Ty89u9XQCAosQHAFCU+AAAihIfAEBR4gMAKKriu9oCH+5UrvhwpQYwkFn5AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAihIfAEBR4gMAKEp8AABFiQ8AoCjxAQAUJT4AgKLEBwBQlPgAAIoa0tsDAL3vwhWPfeQ++9YsKDAJMBBY+QAAihIfAEBR4gMAKEp8AABFiQ8AoChXuwCn5FSuiAE4FVY+AICixAcAUJT4AACKEh8AQFHiAwAoytUu0AtcOQIMZFY+AICixAcAUJT4AACKEh8AQFHiAwAoytUuUAFXqfQ/p/Ke7luzoMAkMHBY+QAAihIfAEBR4gMAKEp8AABFiQ8AoCjxAQAUJT4AgKLOKD7WrFmTQYMGZdmyZZ3bjh07loaGhowdOzYjR47MokWL0tLScqZzAgD9xGnHx3PPPZf77rsv06dP77J9+fLlefTRR/PII49kx44dOXToUK6//vozHhQA6B9OKz7eeeed3HTTTXnggQdy3nnndW5vbW3Ngw8+mO9973u59tprM3PmzGzatCn/9V//lV27dnXb0ABA33Va8dHQ0JAFCxZk3rx5Xbbv3r07J06c6LJ96tSpmTRpUnbu3HlmkwIA/ULF93bZsmVLXnjhhTz33HPve665uTnDhg3L6NGju2yvra1Nc3PzBx6vvb097e3tnV+3tbVVOhIA0IdUtPJx4MCBfO1rX8sPfvCDDB8+vFsGWL16dWpqajofEydO7JbjAgBnp4riY/fu3Tl8+HBmzJiRIUOGZMiQIdmxY0fuvffeDBkyJLW1tTl+/HiOHDnS5ftaWlpSV1f3gcdsampKa2tr5+PAgQOn/Q8DAJz9Kvq1y2c+85m89NJLXbYtXrw4U6dOzR133JGJEydm6NCh2b59exYtWpQk2bNnT/bv35/6+voPPGZVVVWqqqpOc3wAoK+pKD5GjRqVSy65pMu2c889N2PHju3cvmTJkjQ2NmbMmDGprq7O0qVLU19fnyuuuKL7pgYA+qyKTzj9KPfcc08GDx6cRYsWpb29PfPnz8/69eu7+2UAgD7qjOPjqaee6vL18OHDs27duqxbt+5MDw0A9EPdvvIBUMKFKx77yH32rVlQYBKgUm4sBwAUJT4AgKLEBwBQlPgAAIoSHwBAUeIDAChKfAAARYkPAKAo8QEAFCU+AICixAcAUJT4AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAihIfAEBR4gMAKEp8AABFDentAQB6yoUrHit2nH1rFnTLa8FAYOUDAChKfAAARYkPAKAo8QEAFCU+AICixAcAUJT4AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAihIfAEBR4gMAKEp8AABFiQ8AoCjxAQAUJT4AgKLEBwBQlPgAAIoSHwBAUeIDACiqovjYsGFDpk+fnurq6lRXV6e+vj6PP/545/PHjh1LQ0NDxo4dm5EjR2bRokVpaWnp9qEBgL6rovi44IILsmbNmuzevTvPP/98rr322nz+85/PL37xiyTJ8uXL8+ijj+aRRx7Jjh07cujQoVx//fU9MjgA0DcN6ujo6DiTA4wZMybf/e5384UvfCHjxo3L5s2b84UvfCFJ8uqrr2batGnZuXNnrrjiilM6XltbW2pqatLa2prq6uozGQ263YUrHuvtEThL7VuzoLdHgF5Vyc/v0z7n4913382WLVty9OjR1NfXZ/fu3Tlx4kTmzZvXuc/UqVMzadKk7Ny58w8ep729PW1tbV0eAED/NaTSb3jppZdSX1+fY8eOZeTIkdm6dWsuvvjivPjiixk2bFhGjx7dZf/a2to0Nzf/weOtXr06q1atqnhw6G5WNQDKqHjl45Of/GRefPHFPPPMM7ntttty880355VXXjntAZqamtLa2tr5OHDgwGkfCwA4+1W88jFs2LB84hOfSJLMnDkzzz33XP7pn/4pN954Y44fP54jR450Wf1oaWlJXV3dHzxeVVVVqqqqKp8cAOiTzvhzPk6ePJn29vbMnDkzQ4cOzfbt2zuf27NnT/bv35/6+vozfRkAoJ+oaOWjqakp1113XSZNmpS33347mzdvzlNPPZUnnngiNTU1WbJkSRobGzNmzJhUV1dn6dKlqa+vP+UrXQCA/q+i+Dh8+HD++q//Om+++WZqamoyffr0PPHEE/mLv/iLJMk999yTwYMHZ9GiRWlvb8/8+fOzfv36HhkcAOibzvhzPrqbz/mgJ7iShZ7mcz4Y6Ip8zgcAwOkQHwBAUeIDAChKfAAARYkPAKAo8QEAFCU+AICixAcAUJT4AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAihrS2wMA9AcXrnjsI/fZt2ZBgUng7GflAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAihIfAEBR4gMAKEp8AABFiQ8AoCjxAQAU5cZy9HmnckMvAM4eVj4AgKLEBwBQlPgAAIoSHwBAUeIDAChKfAAARYkPAKAo8QEAFCU+AICixAcAUJT4AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAiqooPlavXp3Zs2dn1KhRGT9+fBYuXJg9e/Z02efYsWNpaGjI2LFjM3LkyCxatCgtLS3dOjQA0HdVFB87duxIQ0NDdu3alSeffDInTpzIZz/72Rw9erRzn+XLl+fRRx/NI488kh07duTQoUO5/vrru31wAKBvGlLJztu2bevy9UMPPZTx48dn9+7d+fSnP53W1tY8+OCD2bx5c6699tokyaZNmzJt2rTs2rUrV1xxRfdNDgD0SWd0zkdra2uSZMyYMUmS3bt358SJE5k3b17nPlOnTs2kSZOyc+fODzxGe3t72traujwAgP7rtOPj5MmTWbZsWebOnZtLLrkkSdLc3Jxhw4Zl9OjRXfatra1Nc3PzBx5n9erVqamp6XxMnDjxdEcCAPqA046PhoaGvPzyy9myZcsZDdDU1JTW1tbOx4EDB87oeADA2a2icz5+6/bbb89Pf/rTPP3007ngggs6t9fV1eX48eM5cuRIl9WPlpaW1NXVfeCxqqqqUlVVdTpjAAB9UEUrHx0dHbn99tuzdevW/Md//EemTJnS5fmZM2dm6NCh2b59e+e2PXv2ZP/+/amvr++eiQGAPq2ilY+GhoZs3rw5P/7xjzNq1KjO8zhqamoyYsSI1NTUZMmSJWlsbMyYMWNSXV2dpUuXpr6+3pUuAECSCuNjw4YNSZKrr766y/ZNmzblK1/5SpLknnvuyeDBg7No0aK0t7dn/vz5Wb9+fbcMCwD0fRXFR0dHx0fuM3z48Kxbty7r1q077aEAgP7LvV0AgKLEBwBQlPgAAIoSHwBAUeIDACjqtD7hFICz34UrHvvIffatWVBgEujKygcAUJT4AACKEh8AQFHiAwAoSnwAAEW52oVecypn4kN/4uoTeI+VDwCgKPEBABQlPgCAosQHAFCU+AAAihIfAEBR4gMAKEp8AABFiQ8AoCjxAQAUJT4AgKLc2wWgD3JvJPoyKx8AQFHiAwAoSnwAAEWJDwCgKPEBABTlaheAs4irWBgIrHwAAEWJDwCgKPEBABQlPgCAosQHAFCUq10ABrBTubpm35oFBSZhILHyAQAUJT4AgKLEBwBQlPgAAIoSHwBAUa52oWLuPQHAmbDyAQAUJT4AgKLEBwBQlPgAAIoSHwBAUeIDAChKfAAARVUcH08//XQ+97nPZcKECRk0aFB+9KMfdXm+o6Mjd955Z84///yMGDEi8+bNy2uvvdZd8wIAfVzF8XH06NFceumlWbdu3Qc+/53vfCf33ntvNm7cmGeeeSbnnntu5s+fn2PHjp3xsABA31fxJ5xed911ue666z7wuY6OjqxduzZ/93d/l89//vNJkn/9139NbW1tfvSjH+VLX/rSmU0LAPR53XrOxxtvvJHm5ubMmzevc1tNTU3mzJmTnTt3dudLAQB9VLfe26W5uTlJUltb22V7bW1t53O/r729Pe3t7Z1ft7W1dedIAMBZptdvLLd69eqsWrWqt8fg/3PTOOD3ncp/F/atWVBgEvqLbv21S11dXZKkpaWly/aWlpbO535fU1NTWltbOx8HDhzozpEAgLNMt8bHlClTUldXl+3bt3dua2tryzPPPJP6+voP/J6qqqpUV1d3eQAA/VfFv3Z55513snfv3s6v33jjjbz44osZM2ZMJk2alGXLluWuu+7KRRddlClTpuRb3/pWJkyYkIULF3bn3ABAH1VxfDz//PO55pprOr9ubGxMktx888156KGH8s1vfjNHjx7NrbfemiNHjuSqq67Ktm3bMnz48O6bGgDoswZ1dHR09PYQv6utrS01NTVpbW31K5he4IRT4HQ44ZRKfn67twsAUJT4AACKEh8AQFHiAwAoSnwAAEWJDwCgKPEBABQlPgCAosQHAFCU+AAAihIfAEBR4gMAKKriu9pS3qnc7M1NnQDoK6x8AABFiQ8AoCjxAQAUJT4AgKLEBwBQlKtd+olTuSIGAM4GVj4AgKLEBwBQlPgAAIoSHwBAUeIDACjK1S49yD1ZACrjv5sDg5UPAKAo8QEAFCU+AICixAcAUJT4AACKcrULAP2Oq2bOblY+AICixAcAUJT4AACKEh8AQFHiAwAoytUuAJwxV5dQCSsfAEBR4gMAKEp8AABFiQ8AoCjxAQAU5WoXAIo4lStiSh6nu17LVTyVs/IBABQlPgCAosQHAFCU+AAAihIfAEBRgzo6Ojp6e4jf1dbWlpqamrS2tqa6urrbj1/yLGkASMpeEdNbV+hU8vPbygcAUFSPxce6dety4YUXZvjw4ZkzZ06effbZnnopAKAP6ZH4+OEPf5jGxsasXLkyL7zwQi699NLMnz8/hw8f7omXAwD6kB6Jj+9973u55ZZbsnjx4lx88cXZuHFj/uiP/ij/8i//0hMvBwD0Id3+8erHjx/P7t2709TU1Llt8ODBmTdvXnbu3Pm+/dvb29Pe3t75dWtra5L3TlzpCSfbf9MjxwWAP6SnfqZ9kFP5OdcT8/z2mKdyHUu3x8evfvWrvPvuu6mtre2yvba2Nq+++ur79l+9enVWrVr1vu0TJ07s7tEAoFfUrO3tCbrqyXnefvvt1NTUfOg+vX5juaampjQ2NnZ+ffLkyfzf//1fxo4dm0GDBvXiZP1PW1tbJk6cmAMHDvTIZcycGu/D2cH7cHbwPpwduuN96OjoyNtvv50JEyZ85L7dHh8f+9jHcs4556SlpaXL9paWltTV1b1v/6qqqlRVVXXZNnr06O4ei99RXV3tL/lZwPtwdvA+nB28D2eHM30fPmrF47e6/YTTYcOGZebMmdm+fXvntpMnT2b79u2pr6/v7pcDAPqYHvm1S2NjY26++ebMmjUrn/rUp7J27docPXo0ixcv7omXAwD6kB6JjxtvvDFvvfVW7rzzzjQ3N+eyyy7Ltm3b3ncSKmVVVVVl5cqV7/s1F2V5H84O3oezg/fh7FD6fTjr7u0CAPRv7u0CABQlPgCAosQHAFCU+AAAihIfA9S+ffuyZMmSTJkyJSNGjMgf//EfZ+XKlTl+/Hhvj9bvrVu3LhdeeGGGDx+eOXPm5Nlnn+3tkQaU1atXZ/bs2Rk1alTGjx+fhQsXZs+ePb091oC3Zs2aDBo0KMuWLevtUQacgwcP5q/+6q8yduzYjBgxIn/2Z3+W559/vkdfU3wMUK+++mpOnjyZ++67L7/4xS9yzz33ZOPGjfnbv/3b3h6tX/vhD3+YxsbGrFy5Mi+88EIuvfTSzJ8/P4cPH+7t0QaMHTt2pKGhIbt27cqTTz6ZEydO5LOf/WyOHj3a26MNWM8991zuu+++TJ8+vbdHGXB+/etfZ+7cuRk6dGgef/zxvPLKK/nHf/zHnHfeeT36ui61pdN3v/vdbNiwIa+//npvj9JvzZkzJ7Nnz873v//9JO99+u/EiROzdOnSrFixopenG5jeeuutjB8/Pjt27MinP/3p3h5nwHnnnXcyY8aMrF+/PnfddVcuu+yyrF27trfHGjBWrFiRn//85/nZz35W9HWtfNCptbU1Y8aM6e0x+q3jx49n9+7dmTdvXue2wYMHZ968edm5c2cvTjawtba2Jol/93tJQ0NDFixY0OXvBeX85Cc/yaxZs3LDDTdk/Pjxufzyy/PAAw/0+OuKD5Ike/fuzT//8z/nb/7mb3p7lH7rV7/6Vd599933fdJvbW1tmpube2mqge3kyZNZtmxZ5s6dm0suuaS3xxlwtmzZkhdeeCGrV6/u7VEGrNdffz0bNmzIRRddlCeeeCK33XZbvvrVr+bhhx/u0dcVH/3MihUrMmjQoA99vPrqq12+5+DBg/nLv/zL3HDDDbnlllt6aXIor6GhIS+//HK2bNnS26MMOAcOHMjXvva1/OAHP8jw4cN7e5wB6+TJk5kxY0buvvvuXH755bn11ltzyy23ZOPGjT36uj1ybxd6z9e//vV85Stf+dB9Pv7xj3f++dChQ7nmmmty5ZVX5v777+/h6Qa2j33sYznnnHPS0tLSZXtLS0vq6up6aaqB6/bbb89Pf/rTPP3007ngggt6e5wBZ/fu3Tl8+HBmzJjRue3dd9/N008/ne9///tpb2/POeec04sTDgznn39+Lr744i7bpk2bln//93/v0dcVH/3MuHHjMm7cuFPa9+DBg7nmmmsyc+bMbNq0KYMHWwjrScOGDcvMmTOzffv2LFy4MMl7/9exffv23H777b073ADS0dGRpUuXZuvWrXnqqacyZcqU3h5pQPrMZz6Tl156qcu2xYsXZ+rUqbnjjjuERyFz585936Xmv/zlLzN58uQefV3xMUAdPHgwV199dSZPnpx/+Id/yFtvvdX5nP8L7zmNjY25+eabM2vWrHzqU5/K2rVrc/To0SxevLi3RxswGhoasnnz5vz4xz/OqFGjOs+3qampyYgRI3p5uoFj1KhR7zvP5txzz83YsWOdf1PQ8uXLc+WVV+buu+/OF7/4xTz77LO5//77e3wlXHwMUE8++WT27t2bvXv3vm/J2dXXPefGG2/MW2+9lTvvvDPNzc257LLLsm3btvedhErP2bBhQ5Lk6quv7rJ906ZNH/krS+hvZs+ena1bt6apqSnf/va3M2XKlKxduzY33XRTj76uz/kAAIryS34AoCjxAQAUJT4AgKLEBwBQlPgAAIoSHwBAUeIDAChKfAAARYkPAKAo8QEAFCU+AICixAcAUNT/A1y7G0SkGKJrAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import torch\n",
    "from typing import Optional, Dict\n",
    "from dataclasses import dataclass\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "@dataclass\n",
    "class SimulationConfig:\n",
    "    n_markers: int = 1000\n",
    "    pop_size: int = 1000\n",
    "    h2: float = 1.0\n",
    "    selection_intensity: float = 0.2\n",
    "    \n",
    "class Genome:\n",
    "    def __init__(self, n_markers: int):\n",
    "        self.ploidy: int = 2\n",
    "        self.n_markers: int = n_markers\n",
    "\n",
    "    def __repr__(self) -> str:\n",
    "        return f\"Genome(ploidy={self.ploidy}, n_markers={self.n_markers})\"\n",
    "\n",
    "class Population:\n",
    "    def __init__(self, pop_size: int, genome: Genome, device: torch.device):\n",
    "        self.pop_size: int = pop_size\n",
    "        self.genome: Genome = genome\n",
    "        self.device: torch.device = device\n",
    "        self.haplotypes: torch.Tensor = torch.randint(0, 2, (pop_size, genome.ploidy, genome.n_markers), \n",
    "                                                      dtype=torch.float32, device=self.device)\n",
    "\n",
    "    def to(self, device: torch.device) -> 'Population':\n",
    "        self.device = device\n",
    "        self.haplotypes = self.haplotypes.to(device)\n",
    "        return self\n",
    "\n",
    "    def __repr__(self) -> str:\n",
    "        return f\"Population(pop_size={self.pop_size}, genome={self.genome}, device={self.device})\"\n",
    "\n",
    "class Trait:\n",
    "    def __init__(self, genome: Genome, population: Population, target_mean: float = 0.0, target_variance: float = 1.0):\n",
    "        self.genome: Genome = genome\n",
    "        self.device: torch.device = population.device\n",
    "        self.target_mean: float = target_mean\n",
    "        self.target_variance: float = target_variance\n",
    "\n",
    "        raw_effects = torch.randn(genome.n_markers, device=self.device)\n",
    "        centered_effects = raw_effects - raw_effects.mean()\n",
    "\n",
    "        dosages = population.haplotypes.sum(dim=1)\n",
    "        founder_values = torch.einsum('ij,j->i', dosages, centered_effects)\n",
    "        founder_mean = founder_values.mean()\n",
    "        founder_var = founder_values.var()\n",
    "\n",
    "        scaling_factor = torch.sqrt(self.target_variance / founder_var)\n",
    "        self.effects: torch.Tensor = centered_effects * scaling_factor\n",
    "        self.intercept: torch.Tensor = (torch.tensor(self.target_mean, device=self.device) - founder_mean).detach()\n",
    "\n",
    "    def to(self, device: torch.device) -> 'Trait':\n",
    "        self.device = device\n",
    "        self.effects = self.effects.to(device)\n",
    "        self.intercept = self.intercept.to(device)\n",
    "        return self\n",
    "\n",
    "    def __repr__(self) -> str:\n",
    "        return f\"Trait(target_mean={self.target_mean}, target_variance={self.target_variance}, device={self.device})\"\n",
    "\n",
    "class SimOps:\n",
    "    @staticmethod\n",
    "    def score_population(population: Population, trait: Trait, h2: float = 1.0) -> torch.Tensor:\n",
    "        dosages = population.haplotypes.sum(dim=1)\n",
    "        breeding_values = torch.einsum('ij,j->i', dosages, trait.effects) + trait.intercept\n",
    "        \n",
    "        bv_var = breeding_values.var()\n",
    "        if bv_var == 0 or h2 >= 1:\n",
    "            return breeding_values\n",
    "\n",
    "        env_variance = (1 - h2) / h2 * bv_var.item()\n",
    "        env_std = torch.sqrt(torch.tensor(env_variance, device=population.device))\n",
    "        env_effects = torch.randn_like(breeding_values) * env_std\n",
    "        return breeding_values + env_effects\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def truncation_selection(population: Population, phenotypes: torch.Tensor, selection_intensity: float) -> torch.Tensor:\n",
    "        assert 0 < selection_intensity <= 1, \"Selection intensity must be between 0 and 1\"\n",
    "        assert population.haplotypes.shape[0] == phenotypes.shape[0], \"Mismatch between population size and phenotypes\"\n",
    "        \n",
    "        num_select = max(1, int(selection_intensity * population.pop_size))\n",
    "        _, top_indices = torch.topk(phenotypes, num_select)\n",
    "        return population.haplotypes[top_indices]\n",
    "\n",
    "    @staticmethod\n",
    "    def meiosis(selected_haplotypes: torch.Tensor, num_crossovers: int = 1, num_gametes_per_parent: int = 1) -> torch.Tensor:\n",
    "        num_parents, ploidy, num_markers = selected_haplotypes.shape\n",
    "        \n",
    "        # Repeat each parent's haplotypes num_gametes_per_parent times\n",
    "        expanded_haplotypes = selected_haplotypes.repeat_interleave(num_gametes_per_parent, dim=0)\n",
    "        \n",
    "        # The rest of the function remains largely the same, but operates on the expanded haplotypes\n",
    "        total_gametes = num_parents * num_gametes_per_parent\n",
    "        \n",
    "        crossover_points = torch.randint(1, num_markers, (total_gametes, num_crossovers), device=selected_haplotypes.device)\n",
    "        crossover_points, _ = torch.sort(crossover_points, dim=1)\n",
    "        \n",
    "        crossover_mask = torch.zeros((total_gametes, num_markers), dtype=torch.bool, device=selected_haplotypes.device)\n",
    "        crossover_mask.scatter_(1, crossover_points, 1)\n",
    "        crossover_mask = torch.cumsum(crossover_mask, dim=1) % 2 == 1\n",
    "        \n",
    "        crossover_mask = crossover_mask.unsqueeze(1).expand(-1, ploidy, -1)\n",
    "        \n",
    "        start_chromosome = torch.randint(0, ploidy, (total_gametes, 1), device=selected_haplotypes.device)\n",
    "        start_mask = start_chromosome.unsqueeze(-1).expand(-1, -1, num_markers)\n",
    "        \n",
    "        final_mask = crossover_mask ^ start_mask.bool()\n",
    "        \n",
    "        offspring_haplotypes = torch.where(final_mask, expanded_haplotypes, expanded_haplotypes.roll(shifts=1, dims=1))\n",
    "        \n",
    "        # Return only the first haplotype for each meiosis event\n",
    "        return offspring_haplotypes[:, 0, :]\n",
    "    @staticmethod\n",
    "    def check_cuda(tensor: torch.Tensor, name: str) -> None:\n",
    "        print(f\"{name} is on: {tensor.device}\")\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def random_cross(gamete_tensor: torch.Tensor, total_crosses: int) -> torch.Tensor:\n",
    "        num_gametes, n_markers = gamete_tensor.shape\n",
    "        \n",
    "        # Assert that total_crosses is less than half the number of gametes\n",
    "        assert total_crosses <= num_gametes // 2, f\"Total crosses ({total_crosses}) must be less than or equal to half the number of gametes ({num_gametes // 2})\"\n",
    "        \n",
    "        # Randomly select gametes for crossing\n",
    "        gamete_indices = torch.randperm(num_gametes, device=gamete_tensor.device)\n",
    "        parent1_indices = gamete_indices[:total_crosses]\n",
    "        parent2_indices = gamete_indices[total_crosses:2*total_crosses]\n",
    "        \n",
    "        # Create the new population haplotype tensor\n",
    "        new_population = torch.stack([\n",
    "            gamete_tensor[parent1_indices],\n",
    "            gamete_tensor[parent2_indices]\n",
    "        ], dim=1)\n",
    "        \n",
    "        return new_population\n",
    "\n",
    "##################\n",
    "\n",
    "class SimParams:\n",
    "    def __init__(self, config: SimulationConfig):\n",
    "        self.config = config\n",
    "        self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "        self.genome = Genome(config.n_markers)\n",
    "        self.population = Population(config.pop_size, self.genome, self.device)\n",
    "        self.trait = Trait(self.genome, self.population)\n",
    "\n",
    "\n",
    "    def score_population(self, h2: Optional[float] = None) -> torch.Tensor:\n",
    "        h2 = h2 if h2 is not None else self.config.h2\n",
    "        return SimOps.score_population(self.population, self.trait, h2)\n",
    "\n",
    "    def truncation_selection(self, selection_intensity: Optional[float] = None) -> torch.Tensor:\n",
    "        selection_intensity = selection_intensity if selection_intensity is not None else self.config.selection_intensity\n",
    "        phenotypes = self.score_population()\n",
    "        return SimOps.truncation_selection(self.population, phenotypes, selection_intensity)\n",
    "\n",
    "    def check_device(self) -> None:\n",
    "        SimOps.check_cuda(self.population.haplotypes, \"Population haplotypes\")\n",
    "        SimOps.check_cuda(self.trait.effects, \"Trait effects\")\n",
    "        SimOps.check_cuda(self.trait.intercept, \"Trait intercept\")\n",
    "\n",
    "    def to(self, device: torch.device) -> 'SimParams':\n",
    "        self.device = device\n",
    "        self.population = self.population.to(device)\n",
    "        self.trait = self.trait.to(device)\n",
    "        return self\n",
    "    def perform_meiosis(self, selected_haplotypes: torch.Tensor, num_crossovers: int = 1, num_gametes_per_parent: int = 1) -> torch.Tensor:\n",
    "        return SimOps.meiosis(selected_haplotypes, num_crossovers, num_gametes_per_parent)\n",
    "\n",
    "    def create_next_generation(self, gametes: torch.Tensor, total_crosses: int) -> None:\n",
    "        new_population_haplotypes = SimOps.random_cross(gametes, total_crosses)\n",
    "        self.population.haplotypes = new_population_haplotypes\n",
    "        self.population.pop_size = total_crosses\n",
    "\n",
    "\n",
    "########### SIM PARAMS  ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "\n",
    "\n",
    "\n",
    "# Usage\n",
    "config = SimulationConfig(\n",
    "    n_markers=10000,\n",
    "    pop_size=1000,\n",
    "    h2=0.5,\n",
    "    selection_intensity=0.2)\n",
    "\n",
    "SP = SimParams(config)\n",
    "SP.to(SP.device)  # Ensure all tensors are on the correct device\n",
    "\n",
    "scores = SP.score_population()\n",
    "\n",
    "plt.hist(SP.score_population().detach().cpu().numpy(), bins=50)\n",
    "SP.check_device()\n",
    "SimOps.check_cuda(scores, \"Population scores\")\n",
    "\n",
    "print(f\"Variance of scores: {scores.var()}\")\n",
    "\n",
    "selected_haplotypes = SP.truncation_selection()\n",
    "SimOps.check_cuda(selected_haplotypes, \"Selected haplotypes\")\n",
    "\n",
    "print(f\"Original population shape: {SP.population.haplotypes.shape}\")\n",
    "print(f\"Selected haplotypes shape: {selected_haplotypes.shape}\")\n",
    "\n",
    "# After performing meiosis and generating gametes\n",
    "gametes = SP.perform_meiosis(selected_haplotypes, num_crossovers=3, num_gametes_per_parent=5)\n",
    "SimOps.check_cuda(gametes, \"Gametes\")\n",
    "print(f\"Gametes shape: {gametes.shape}\")\n",
    "\n",
    "# Perform random crosses to create the next generation\n",
    "total_crosses = min(SP.config.pop_size, gametes.shape[0] // 2)\n",
    "SP.create_next_generation(gametes, total_crosses)\n",
    "\n",
    "print(f\"New population shape: {SP.population.haplotypes.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f3d47a96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.57 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "scores_h2_05 = SP.score_population(h2=0.5)  # Equal genetic and environmental variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e83674cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.7 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "gametes = SP.perform_meiosis(selected_haplotypes, num_crossovers=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96b48133",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.63 ms ± 172 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "total_crosses = min(SP.config.pop_size, gametes.shape[0] // 2)\n",
    "SP.create_next_generation(gametes, total_crosses)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
